{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up environment\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "import sgkit as sg\n",
    "import json\n",
    "import hashlib\n",
    "import itertools as it\n",
    "\n",
    "\n",
    "import allel; print('scikit-allel', allel.__version__)\n",
    "\n",
    "#px config\n",
    "config = {\n",
    "  'toImageButtonOptions': {\n",
    "    'format': 'png', # one of png, svg, jpeg, webp\n",
    "    'filename': 'custom_image',\n",
    "    'height': 500,\n",
    "    'width': 700,\n",
    "    'scale':6 # Multiply title/legend/axis/canvas sizes by this factor\n",
    "  }\n",
    "}\n",
    "\n",
    "#palettes\n",
    "pop_code_cols = {\n",
    "    'SAE' : '#6a3d9a', #dark purple\n",
    "    'SAR' : '#cab2d6', #ligher purple\n",
    "    'INB' : '#96172e', #darkred\n",
    "    'INM' : '#f03e5e', #lightred\n",
    "    'APA' : '#ff7f00', #orange\n",
    "    'IRN' : '#C2907A', #not sure yet\n",
    "    'DJI' : '#507d2a', #sap green\n",
    "    'ETW' : '#a6cee3',#cerulean\n",
    "    'ETB' : '#007272', #cobalt turq\n",
    "    'ETS' : '#33a02c',#green\n",
    "    'SUD' : '#fccf86',#ochre\n",
    "    'YEM' : '#CC7722'#pinkish\n",
    "}\n",
    "\n",
    "#load and filter metadata\n",
    "#load and filter metadata\n",
    "df_samples = pd.read_csv('/Users/dennistpw/Projects/AsGARD/metadata/cease_combinedmetadata.20250212.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up population dictionary to store per-pop allele counts\n",
    "\n",
    "# Initialize an empty dictionary to store row indices for each level\n",
    "pop_dict = {}\n",
    "\n",
    "# Iterate through unique levels in the 'factor_column'\n",
    "for pop in df_samples['pop_code'].unique():\n",
    "    pop_dict[pop] = df_samples.index[df_samples['pop_code'] == pop].tolist()\n",
    "\n",
    "# Remove dictionary entries with fewer than 5 values\n",
    "pop_dict = {key: value for key, value in pop_dict.items() if len(value) >= 5}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper functions\n",
    "def hash_params(*args, **kwargs):\n",
    "    \"\"\"Helper function to hash analysis parameters.\"\"\"\n",
    "    o = {\n",
    "        'args': args,\n",
    "        'kwargs': kwargs\n",
    "    }\n",
    "    s = json.dumps(o, sort_keys=True).encode()\n",
    "    h = hashlib.md5(s).hexdigest()\n",
    "    return h\n",
    "\n",
    "#function for getting allele counts\n",
    "def snp_allele_counts(\n",
    "    #gets all allele counts for a given chrom\n",
    "\n",
    "        ds=None,\n",
    "        sample_query=None,\n",
    "        sample_list=None,\n",
    "        min_ac=1\n",
    "        ):\n",
    "     \n",
    "    if sample_query:\n",
    "        loc_samples_a = df_samples.eval(sample_query).values\n",
    "        ds = ds.isel(samples=loc_samples_a)\n",
    "    elif sample_list:\n",
    "        loc_samples_a = df_samples['sample_id'].isin(sample_list)\n",
    "        ds = ds.isel(samples=loc_samples_a)\n",
    "\n",
    "    ac = allel.GenotypeArray(ds.call_genotype.values).count_alleles()\n",
    "\n",
    "    #need to make maf filter\n",
    "    #if min_ac\n",
    "\n",
    "    return ac\n",
    "\n",
    "# Function to run Fst genome scan\n",
    "def do_fst_scan(\n",
    "          chrom=None,\n",
    "          sample_query_a=None, \n",
    "          sample_query_b=None,\n",
    "          sample_list_a = None,\n",
    "          sample_list_b=None,\n",
    "          winsize=10000,\n",
    "          analysis_name = 'fst',\n",
    "          results_dir=None\n",
    "    ):\n",
    "            # construct a key to save the results under\n",
    "    results_key = hash_params(\n",
    "        chrom=chrom,\n",
    "        sample_query_a=sample_query_a, \n",
    "        sample_query_b=sample_query_b,\n",
    "        sample_list_a = sample_list_a,\n",
    "        sample_list_b=sample_list_b, \n",
    "        winsize=winsize,\n",
    "        analysis_name = analysis_name,\n",
    "        results_dir=results_dir\n",
    "        )\n",
    "\n",
    "        # define paths for results files\n",
    "    data_path = f'{results_dir}/{results_key}-fst.csv'\n",
    "\n",
    "    try:\n",
    "        # try to load previously generated results\n",
    "        data = pd.read_csv(data_path)\n",
    "        return data\n",
    "    except FileNotFoundError:\n",
    "        # no previous results available, need to run analysis\n",
    "        print(f'running analysis: {results_key}')\n",
    "\n",
    "    print('setting up inputs')\n",
    "\n",
    "    #load ds\n",
    "    ds = sg.load_dataset(f'/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/combined_cohorts.{chrom}.zarr')\n",
    "\n",
    "    # Get accessible only\n",
    "    print(\"Subsetting to accessible sites only\")\n",
    "    accmask = ds['is_accessible'].compute()\n",
    "    ds = ds.sel(variants=(accmask))\n",
    "\n",
    "    #get allele counts for a\n",
    "    if sample_query_a:\n",
    "        ac1 = snp_allele_counts(ds, sample_query = sample_query_a)\n",
    "    else:\n",
    "        ac1 = snp_allele_counts(ds, sample_list = sample_list_a)\n",
    "\n",
    "    #get ac for b\n",
    "    if sample_query_a:\n",
    "        ac2 = snp_allele_counts(ds, sample_query = sample_query_b)\n",
    "    else:\n",
    "        ac2 = snp_allele_counts(ds, sample_list = sample_list_b)\n",
    "    \n",
    "    #get pos\n",
    "    pos = ds.variant_position.values\n",
    "\n",
    "    print(\"computing Fst\")\n",
    "    fst = allel.moving_hudson_fst(ac1, ac2, size=winsize)\n",
    "    # Sometimes Fst can be very slightly below zero, clip for simplicity.\n",
    "    fst = np.clip(fst, a_min=0, a_max=1)\n",
    "    x = allel.moving_statistic(pos, statistic=np.mean, size=winsize)\n",
    "\n",
    "    # save results\n",
    "    fstdf = pd.DataFrame(\n",
    "            {'chrom' : chrom,\n",
    "            'midpos':x,\n",
    "            'fst':fst})\n",
    "\n",
    "    fstdf.to_csv(data_path, index=False)\n",
    "    print(f'saved results: {results_key}')\n",
    "\n",
    "    return(fstdf)\n",
    "\n",
    "# Function to calculate genomewide average Fst\n",
    "\n",
    "def do_fst_av(\n",
    "          chrom=None,\n",
    "          sample_query_a=None, \n",
    "          sample_query_b=None,\n",
    "          sample_list_a = None,\n",
    "          sample_list_b=None,\n",
    "          winsize=10000,\n",
    "          analysis_name = 'fst_average',\n",
    "          results_dir=None\n",
    "    ):\n",
    "            # construct a key to save the results under\n",
    "    results_key = hash_params(\n",
    "        chrom=chrom,\n",
    "        sample_query_a=sample_query_a, \n",
    "        sample_query_b=sample_query_b,\n",
    "        sample_list_a = sample_list_a,\n",
    "        sample_list_b=sample_list_b, \n",
    "        winsize=winsize,\n",
    "        analysis_name = analysis_name,\n",
    "        results_dir=results_dir\n",
    "        )\n",
    "\n",
    "        # define paths for results files\n",
    "    data_path = f'{results_dir}/{results_key}-fst.csv'\n",
    "\n",
    "    try:\n",
    "        # try to load previously generated results\n",
    "        data = pd.read_csv(data_path)\n",
    "        return data\n",
    "    except FileNotFoundError:\n",
    "        # no previous results available, need to run analysis\n",
    "        print(f'running analysis: {results_key}')\n",
    "\n",
    "    print('setting up inputs')\n",
    "\n",
    "    #load ds\n",
    "    ds = sg.load_dataset(f'/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/combined_cohorts.{chrom}.zarr')\n",
    "\n",
    "    #get allele counts for a\n",
    "    if sample_query_a:\n",
    "        ac1 = snp_allele_counts(ds, sample_query = sample_query_a)\n",
    "    else:\n",
    "        ac1 = snp_allele_counts(ds, sample_list = sample_list_a)\n",
    "\n",
    "    #get ac for b\n",
    "    if sample_query_a:\n",
    "        ac2 = snp_allele_counts(ds, sample_query = sample_query_b)\n",
    "    else:\n",
    "        ac2 = snp_allele_counts(ds, sample_list = sample_list_b)\n",
    "    \n",
    "    #get pos\n",
    "    pos = ds.variant_position.values\n",
    "\n",
    "    print(\"allele counts and pos loaded\")\n",
    "\n",
    "    print(\"computing average Hudson Fst\")\n",
    "    fst = allel.average_hudson_fst(ac1, ac2, blen=winsize)\n",
    "\n",
    "\n",
    "    fdict = {'fst':fst[0],'se':fst[1]}\n",
    "    fstdf = pd.DataFrame([fdict])\n",
    "\n",
    "    fstdf.to_csv(data_path, index=False)\n",
    "    print(f'saved results: {results_key}')\n",
    "\n",
    "    return(fstdf)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do genomescan (lots of output)\n",
    "\n",
    "fstlist = []\n",
    "winsize=10000\n",
    "output = '/Users/dennistpw/Projects/AsGARD/data/fst_20240712'\n",
    "\n",
    "for chrom in ['CM023248','CM023249','CM023250']:\n",
    "    for popa, popb in it.combinations(df_samples.pop_code.unique(),2):\n",
    "        \n",
    "        fst_df = do_fst_scan(\n",
    "            chrom = chrom,\n",
    "            sample_query_a = f'pop_code == \"{popa}\"',\n",
    "            sample_query_b = f'pop_code == \"{popb}\"',\n",
    "            results_dir = output,\n",
    "            analysis_name=f'{chrom}.{winsize}.{popa}.{popb}'\n",
    "        )\n",
    "\n",
    "        fst_df['popa'] = popa\n",
    "        fst_df['popb'] = popb\n",
    "\n",
    "        fstlist.append(fst_df)\n",
    "\n",
    "    \n",
    "fst_bigdf  = pd.concat(fstlist)\n",
    "\n",
    "fst_bigdf['comp'] = fst_bigdf['popa'] +'_' + fst_bigdf['popb']\n",
    "\n",
    "from matplotlib.gridspec import GridSpec\n",
    "\n",
    "max_values = fst_bigdf.groupby('chrom')['midpos'].max()\n",
    "total_max = max_values.sum()\n",
    "column_widths = (max_values / total_max).values\n",
    "col_var_levels = max_values.index\n",
    "    \n",
    "cyp6_region = {'CM023248': {'x_min': 67473117, 'x_max': 67501071, 'y_min': 0, 'y_max': 1}}\n",
    "ace1_region = {'CM023248': {'x_min': 60916071, 'x_max': 60917000, 'y_min': 0, 'y_max': 1}}\n",
    "vgsc_region = {'CM023249': {'x_min': 42817709, 'x_max': 42817800, 'y_min': 0, 'y_max': 1}}\n",
    "gste_region = {'CM023249': {'x_min': 70572788, 'x_max': 70584603, 'y_min': 0, 'y_max': 1}}\n",
    "carboxylesterase_region = {'CM023249':{'x_min' : 18816120, 'x_max' :18816120, 'y_min': 0, 'y_max': 1}}\n",
    "cyp9_region = {'CM023250': {'x_min': 9721225, 'x_max': 9722225, 'y_min': 0, 'y_max': 1}}\n",
    "diagk = {'CM023250': {'x_min': 4578144, 'x_max': 4578144, 'y_min': 0, 'y_max': 1}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot genome scan\n",
    "\n",
    "# Get the unique row and column variables\n",
    "row_var_levels = fst_bigdf['comp'].unique()\n",
    "\n",
    "# Initialize the figure and GridSpec\n",
    "fig = plt.figure(figsize=(sum(column_widths) * 14, len(row_var_levels) * 2))  # Plots half as high\n",
    "gs = GridSpec(len(row_var_levels), len(col_var_levels) + 1, width_ratios=list(column_widths) + [0.1])\n",
    "\n",
    "# Create the subplots\n",
    "for row_idx, row_val in enumerate(row_var_levels):\n",
    "    for col_idx, col_val in enumerate(col_var_levels):\n",
    "        ax = fig.add_subplot(gs[row_idx, col_idx])\n",
    "        subset = fst_bigdf[(fst_bigdf['comp'] == row_val) & (fst_bigdf['chrom'] == col_val)]\n",
    "        sns.lineplot(data=subset, x='midpos', y='fst', ax=ax)\n",
    "\n",
    "        ax.fill_between(x=subset['midpos'], y1=subset['fst'], color='blue', alpha=0.1)  # Adjust alpha for transparency if needed\n",
    "\n",
    "        #ax.set_title(f\"{row_val} - {col_val}\")\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "        # Remove y-axis for 2nd and 3rd columns\n",
    "        if col_idx > 0:\n",
    "            ax.set_ylabel('')\n",
    "            ax.yaxis.set_visible(False)\n",
    "\n",
    "        if col_val == 'CM023248':\n",
    "            ax.fill_betweenx([cyp6_region[col_val]['y_min'], cyp6_region[col_val]['y_max']],\n",
    "                            cyp6_region[col_val]['x_min'], cyp6_region[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.3)\n",
    "            ax.fill_betweenx([ace1_region[col_val]['y_min'], ace1_region[col_val]['y_max']],\n",
    "                            ace1_region[col_val]['x_min'], ace1_region[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.3)\n",
    "        elif col_val == 'CM023249':\n",
    "            ax.fill_betweenx([vgsc_region[col_val]['y_min'], vgsc_region[col_val]['y_max']],\n",
    "                            vgsc_region[col_val]['x_min'], vgsc_region[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.3)\n",
    "            ax.fill_betweenx([gste_region[col_val]['y_min'], gste_region[col_val]['y_max']],\n",
    "                            gste_region[col_val]['x_min'], gste_region[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.3)\n",
    "            ax.fill_betweenx([carboxylesterase_region[col_val]['y_min'], carboxylesterase_region[col_val]['y_max']],\n",
    "                            carboxylesterase_region[col_val]['x_min'], carboxylesterase_region[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.3)\n",
    "        elif col_val == 'CM023250':\n",
    "            ax.fill_betweenx([cyp9_region[col_val]['y_min'], cyp9_region[col_val]['y_max']],\n",
    "                            cyp9_region[col_val]['x_min'], cyp9_region[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.3)\n",
    "            ax.fill_betweenx([diagk[col_val]['y_min'], diagk[col_val]['y_max']],\n",
    "                            diagk[col_val]['x_min'], diagk[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.3)\n",
    "            \n",
    "        # Remove x-axis\n",
    "        ax.set_xlabel('')\n",
    "        #ax.xaxis.set_visible(False)\n",
    "\n",
    "        ax.set_ylim(0, 0.8)\n",
    "\n",
    "        \n",
    "    # Add the row title\n",
    "    row_ax = fig.add_subplot(gs[row_idx, -1])\n",
    "    row_ax.text(0.5, 0.5, row_val, va='center', ha='center', fontsize=12, rotation=90, transform=row_ax.transAxes)\n",
    "    row_ax.axis('off')\n",
    "\n",
    "plt.subplots_adjust(wspace=0.01, hspace=0.01)  # Reduce column and row padding\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "fig.savefig('../figures/fstscan.svg')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate average pairwise Fst between cohorts\n",
    "\n",
    "fstlist = []\n",
    "output = '/Users/dennistpw/Projects/AsGARD/data/fst_20240712'\n",
    "\n",
    "for chrom in ['CM023248','CM023249','CM023250']:\n",
    "    for popa, popb in it.combinations(df_samples.pop_code.unique(),2):\n",
    "        \n",
    "        hud_fst_df = do_fst_av(\n",
    "            chrom = chrom,\n",
    "            sample_query_a = f'pop_code == \"{popa}\"',\n",
    "            sample_query_b = f'pop_code == \"{popb}\"',\n",
    "            results_dir = output,\n",
    "            #analysis_name=f'{chrom}.{winsize}.{popa}.{popb}.hudfst'\n",
    "        )\n",
    "\n",
    "        hud_fst_df['popa'] = popa\n",
    "        hud_fst_df['popb'] = popb\n",
    "        hud_fst_df['chrom'] = chrom\n",
    "\n",
    "        fstlist.append(hud_fst_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot heatmap\n",
    "\n",
    "# Make ordered df for plotting as a heatmap\n",
    "\n",
    "# Ordered list of cohorts\n",
    "order = ['YEM',\n",
    " 'SUD',\n",
    " 'ETS',\n",
    " 'ETB',\n",
    " 'ETW',\n",
    " 'DJI',\n",
    " 'INM',\n",
    " 'INB',\n",
    " 'APA',\n",
    " 'IRS',\n",
    " 'IRH',\n",
    " 'SAR',\n",
    " 'SAE']\n",
    "\n",
    "# Make df\n",
    "fst_hud_df = pd.concat(fstlist)\n",
    "sub = fst_hud_df[fst_hud_df['chrom'] == 'CM023248']\n",
    "sub2 = sub.copy()\n",
    "order = list(pop_code_cols.keys())\n",
    "order.reverse()\n",
    "\n",
    "sub2.columns = ['fst', 'se', 'popb', 'popa', 'chrom']\n",
    "subc=pd.concat([sub,sub2])\n",
    "pivot_df = subc.pivot(index=\"popa\", columns=\"popb\", values=\"fst\")\n",
    "\n",
    "#reorder df\n",
    "reordered_df = pivot_df.loc[order, order]\n",
    "\n",
    "# Create a mask for the lower triangle\n",
    "mask = np.triu(np.ones_like(reordered_df, dtype=bool))\n",
    "sns.set_style('white')\n",
    "\n",
    "# Plot the heatmap with the mask\n",
    "plt.figure(figsize=(8, 6))\n",
    "# Plot the heatmap with the mask\n",
    "plt.figure(figsize=(8, 6))\n",
    "hm = sns.heatmap(\n",
    "    reordered_df,\n",
    "    mask=mask,\n",
    "    cmap='viridis',  # Use viridis color palette\n",
    "    annot=True,\n",
    "    cbar=True,\n",
    "    fmt=\".2f\",\n",
    "    annot_kws={\"size\": 12.5}\n",
    ")\n",
    "\n",
    "# Customize axis labels\n",
    "hm.set_xlabel('Population A')\n",
    "hm.set_ylabel('Population B')\n",
    "plt.savefig('../figures/fst_heatmap.svg')\n",
    "plt.savefig('../figures/fst_heatmap.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now get pairwise Fst between all sample locations for spatial analaysis\n",
    "\n",
    "# Now, fst between all invasive locations for analysis and modelling in R\n",
    "loc_groups = df_samples.query('country != \"Pakistan\" & country != \"Afghanistan\" & country != \"SaudiArabia\" & country != \"India\"').groupby('location').count()\n",
    "locs = loc_groups['sample_id'][loc_groups['sample_id'] > 5].index.tolist()\n",
    "locs\n",
    "#cheap manual reorder\n",
    "locs = ['Nangarhar','DjiboutiCity','AdenCity','Dubti','Jiga','Modjo','Babile','KebriDehar','Danan','PortSudan','Haiya','AlShukria','EastElglabat','ElZedab','SouthShandi','GeziraIslang','Arkaweet','ElSalamaniaWest','ElMeaileg','Wafara','AlGalaa','Agaja']\n",
    "\n",
    "#Let's quickly calculate Fst here by location, in locs with > 10 samples, to infer potential isoBD\n",
    "fstlist = []\n",
    "winsize=10000\n",
    "output = '/Users/dennistpw/Projects/AsGARD/data/fst_byloc/'\n",
    "\n",
    "for chrom in ['CM023248','CM023249','CM023250']:\n",
    "    for popa, popb in it.combinations(locs,2):\n",
    "        \n",
    "        hud_fst_df = do_fst_av(\n",
    "            chrom = chrom,\n",
    "            sample_query_a = f'location == \"{popa}\"',\n",
    "            sample_query_b = f'location == \"{popb}\"',\n",
    "            results_dir = output,\n",
    "            analysis_name=f'{chrom}.{winsize}.{popa}.{popb}.hudfst'\n",
    "        )\n",
    "\n",
    "        hud_fst_df['loca'] = popa\n",
    "        hud_fst_df['locb'] = popb\n",
    "        hud_fst_df['chrom'] = chrom\n",
    "\n",
    "        fstlist.append(hud_fst_df)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cease_local",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
