{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anjl\n",
    "import numba\n",
    "import sgkit as sg\n",
    "import allel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#chrom dict\n",
    "scaflens={'CM023248' : 93706023,\n",
    "'CM023249' : 88747589,\n",
    "'CM023250' : 22713616}\n",
    "\n",
    "df_samples = pd.read_csv('/Users/dennistpw/Projects/AsGARD/metadata/cease_combinedmetadata_noqc.20250212.csv')\n",
    "\n",
    "#palettes\n",
    "#palettes\n",
    "pop_code_cols = {\n",
    "    'APA' : '#ff7f00', #orange\n",
    "    'SAE' : '#6a3d9a', #dark purple\n",
    "    'SAR' : '#cab2d6', #ligher purple\n",
    "    'IRS' : '#c27a88', #not sure yet\n",
    "    'IRH' : '#c57fc9', #not sure yet\n",
    "    'INB' : '#96172e', #darkred\n",
    "    'INM' : '#f03e5e', #lightred\n",
    "    'DJI' : '#507d2a', #sap green\n",
    "    'ETB' : '#007272', #cobalt turq\n",
    "    'ETS' : '#33a02c',#green\n",
    "    'ETW' : '#a6cee3',#cerulean\n",
    "    'SUD' : '#fccf86',#ochre\n",
    "    'YEM' : '#CC7722'#pinkish\n",
    "}\n",
    "\n",
    "contigs_dict = {'CM023248':0,\n",
    "                'CM023249':1,\n",
    "                'CM023250':2}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define stevegen500 functions\n",
    "def select_random_genos(\n",
    "                    ds, \n",
    "                    n_snps=100_000):\n",
    "                     #selects given number of genos at random\n",
    "                    keep_indices = np.random.choice(ds.call_genotype.shape[0], n_snps, replace=False)\n",
    "                    keep_indices.sort()\n",
    "                    thinned_callset = ds.isel(variants=~keep_indices)\n",
    "                    return(thinned_callset)\n",
    "\n",
    "def load_geno_ds(chrom, \n",
    "                 df_samples,\n",
    "                sample_query=None, \n",
    "                n_snps=None, \n",
    "                sample_list=None, \n",
    "                start=None, \n",
    "                end=None, \n",
    "                min_minor_ac=None,\n",
    "                acconly=True\n",
    "                ):\n",
    "                 # load sample metadata\n",
    "     #load ds\n",
    "     ds = sg.load_dataset(f'/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/combined_cohorts.{chrom}.zarr')\n",
    "\n",
    "     if sample_query:\n",
    "          # locate selected samples\n",
    "        loc_samples = df_samples.query(sample_query).index\n",
    "        df_samples = df_samples.loc[loc_samples, :]\n",
    "        ds = ds.isel(samples=loc_samples)\n",
    "     elif sample_list:\n",
    "          loc_samples = df_samples['sample_id'].isin(sample_list)\n",
    "          df_samples = df_samples.loc[loc_samples, :]\n",
    "          ds = ds.isel(samples=loc_samples)\n",
    "     else:\n",
    "          pass\n",
    "          \n",
    "     \n",
    "     #if minmaf is specified, select minmaf  \n",
    "     if min_minor_ac:   \n",
    "        print(f'subsetting to segregating sites')\n",
    "        ac = allel.GenotypeArray(ds['call_genotype']).count_alleles()\n",
    "        macbool = ac[:,1] >= min_minor_ac\n",
    "        print(f'selected {np.sum(macbool)} sites with a min mac > {min_minor_ac}')\n",
    "        ds_analysis = ds.sel(variants=(macbool))\n",
    "     else:\n",
    "        ds_analysis = ds\n",
    "\n",
    "     #get accessible only\n",
    "     if acconly:\n",
    "        print('subsetting to accessible sites only')\n",
    "        accmask = ds_analysis['is_accessible'].compute()\n",
    "        ds_analysis = ds_analysis.sel(variants=(accmask))\n",
    "     else:\n",
    "            pass\n",
    "     \n",
    "    #if numgenos is set, subset \n",
    "     if n_snps:\n",
    "          ds_analysis = select_random_genos(ds_analysis, n_snps=n_snps)\n",
    "     elif start:\n",
    "        print(f\"subsetting haps to range {chrom}:{start}-{end}\")\n",
    "        ds_analysis = ds_analysis.set_index(variants=(\"variant_contig\", \"variant_position\")).sel(variants=(0, slice(start,end)))\n",
    "     else:\n",
    "            pass\n",
    "\n",
    "     #return completed ds\n",
    "     return(df_samples, ds_analysis)\n",
    "\n",
    "\n",
    "# Define helper functions\n",
    "@numba.njit\n",
    "def square_to_condensed(i, j, n):\n",
    "    \"\"\"Convert distance matrix coordinates from square form (i, j) to condensed form.\"\"\"\n",
    "\n",
    "    assert i != j, \"no diagonal elements in condensed matrix\"\n",
    "    if i < j:\n",
    "        i, j = j, i\n",
    "    return n * j - j * (j + 1) // 2 + i - 1 - j\n",
    "\n",
    "@numba.njit\n",
    "def genotype_cityblock(x, y):\n",
    "    n_sites = x.shape[0]\n",
    "    distance = np.float32(0)\n",
    "\n",
    "    # Loop over sites.\n",
    "    for i in range(n_sites):\n",
    "        # Compute cityblock distance (absolute difference).\n",
    "        d = np.fabs(x[i] - y[i])\n",
    "\n",
    "        # Accumulate distance for the current pair.\n",
    "        distance += d\n",
    "\n",
    "    return distance\n",
    "\n",
    "@numba.njit(parallel=True)\n",
    "def biallelic_diplotype_pdist(X):\n",
    "    n_samples = X.shape[0]\n",
    "    n_pairs = (n_samples * (n_samples - 1)) // 2\n",
    "    out = np.zeros(n_pairs, dtype=np.float32)\n",
    "\n",
    "    # Loop over samples, first in pair.\n",
    "    for i in range(n_samples):\n",
    "        x = X[i, :]\n",
    "\n",
    "        # Loop over observations again, second in pair.\n",
    "        for j in numba.prange(i + 1, n_samples):\n",
    "            y = X[j, :]\n",
    "\n",
    "            # Compute distance for the current pair.\n",
    "            d = genotype_cityblock(x, y)\n",
    "\n",
    "            # Store result for the current pair.\n",
    "            k = square_to_condensed(i, j, n_samples)\n",
    "            out[k] = d\n",
    "\n",
    "    return out\n",
    "\n",
    "\n",
    "def get_pdist( \n",
    "    chrom,\n",
    "    df_samples,\n",
    "    sample_query=None,\n",
    "    min_minor_ac=1,\n",
    "    n_snps=100_000,\n",
    "    sample_list=None):   \n",
    "\n",
    "     # access SNP calls\n",
    "    sample_df, ds = load_geno_ds(\n",
    "          chrom=chrom,\n",
    "          df_samples=df_samples,\n",
    "          n_snps=n_snps,\n",
    "          min_minor_ac=min_minor_ac,\n",
    "          sample_query=sample_query,\n",
    "          sample_list=sample_list,\n",
    "          acconly = True)\n",
    "    \n",
    "   \n",
    "    # access SNP genotypes\n",
    "    gn = allel.GenotypeArray(ds[\"call_genotype\"]).to_n_alt()\n",
    "\n",
    "    X = np.ascontiguousarray(gn.T)\n",
    "\n",
    "    dist = biallelic_diplotype_pdist(X)\n",
    "    np.save(f\"/Users/dennistpw/Projects/AsGARD/data/{chrom}.{n_snps}.{min_minor_ac}.dist.npy\", dist)\n",
    "    return(dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import squareform  # type: ignore\n",
    "\n",
    "\n",
    "def plot_njt(chrom, filename, n_snps):\n",
    "    # Get pairwise distance\n",
    "    dist = get_pdist(chrom=chrom,df_samples=df_samples, n_snps=n_snps)\n",
    "    #Coerce to square matrix\n",
    "    D = squareform(dist)\n",
    "    # Build the NJ tree\n",
    "    progress_options = dict(desc=\"Construct neighbour-joining tree\", leave=False)\n",
    "    Z = anjl.canonical_nj(\n",
    "                    D=D, progress_options=progress_options\n",
    "    )\n",
    "    # Plot\n",
    "    nj_2 = anjl.plot(\n",
    "                Z=Z,\n",
    "                leaf_data=df_samples,\n",
    "                color=\"pop_code\",\n",
    "                color_discrete_map=pop_code_cols,\n",
    "    )\n",
    "    #Save\n",
    "    nj_2.write_image(f\"/Users/dennistpw/Projects/AsGARD/figures/{filename}.svg\")\n",
    "    nj_2.show()\n",
    "    return(Z)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "z3 = plot_njt('CM023248',\n",
    "         filename='njt_chr2',\n",
    "         n_snps=100_000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(z3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_njt('CM023249',\n",
    "         filename='njt_chr3',\n",
    "        n_snps=100_000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_njt('CM023250',\n",
    "         filename='njt_chrX',\n",
    "        n_snps=100_000)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cease_local",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
