{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1d3edc6-0c7a-44b3-8ece-be1901ddc468",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "import sgkit as sg\n",
    "import json\n",
    "import hashlib\n",
    "import plotly.express as px\n",
    "import dask\n",
    "import allel; print('scikit-allel', allel.__version__)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e02fe1d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#px config\n",
    "config = {\n",
    "  'toImageButtonOptions': {\n",
    "    'format': 'png', # one of png, svg, jpeg, webp\n",
    "    'filename': 'custom_image',\n",
    "    'height': 500,\n",
    "    'width': 700,\n",
    "    'scale':6 # Multiply title/legend/axis/canvas sizes by this factor\n",
    "  }\n",
    "}\n",
    "\n",
    "#palettes\n",
    "pop_code_cols = {\n",
    "    'APA' : '#ff7f00', #orange\n",
    "    'SAE' : '#6a3d9a', #dark purple\n",
    "    'SAR' : '#cab2d6', #ligher purple\n",
    "    'IRS' : '#c27a88', #not sure yet\n",
    "    'IRH' : '#c57fc9', #not sure yet\n",
    "    'INB' : '#96172e', #darkred\n",
    "    'INM' : '#f03e5e', #lightred\n",
    "    'DJI' : '#507d2a', #sap green\n",
    "    'ETB' : '#007272', #cobalt turq\n",
    "    'ETS' : '#33a02c',#green\n",
    "    'ETW' : '#a6cee3',#cerulean\n",
    "    'SUD' : '#fccf86',#ochre\n",
    "    'YEM' : '#CC7722'#pinkish\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0faac09-c480-48d5-b4b9-28decac492e1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#define helper functions\n",
    "def select_random_genos(\n",
    "                    ds, \n",
    "                    numgenos):\n",
    "                     #selects given number of genos at random\n",
    "                    keep_indices = np.random.choice(ds.call_genotype.shape[0], numgenos, replace=False)\n",
    "                    keep_indices.sort()\n",
    "                    thinned_callset = ds.isel(variants=~keep_indices)\n",
    "                    return(thinned_callset)\n",
    "\n",
    "\n",
    "#some funcs\n",
    "def plot_variance(evr, **kwargs):\n",
    "    \"\"\"Plot a bar chart showing variance explained by each principal\n",
    "    component.\"\"\"\n",
    "    \n",
    "    # prepare variables\n",
    "    y = evr * 100\n",
    "    x = [str(i+1) for i in range(len(y))]\n",
    "    \n",
    "    # setup plotting options\n",
    "    plot_kwargs = dict(\n",
    "        labels={\n",
    "            'x': 'Principal component',\n",
    "            'y': 'Explained variance (%)',\n",
    "        },\n",
    "        template='simple_white',\n",
    "        width=600,\n",
    "        height=400\n",
    "    )\n",
    "    # apply any user overrides\n",
    "    plot_kwargs.update(kwargs)\n",
    "\n",
    "    # make a bar plot\n",
    "    fig = px.bar(x=x, y=y, **plot_kwargs)\n",
    "    fig.show()\n",
    "    \n",
    "    \n",
    "def jitter(a, f):\n",
    "    r = a.max() - a.min()\n",
    "    return a + f * np.random.uniform(-r, r, a.shape)\n",
    "\n",
    "\n",
    "def plot_coords(\n",
    "    data,\n",
    "    x='PC1',\n",
    "    y='PC2',\n",
    "    colour = 'pop_code',\n",
    "    palette = px.colors.qualitative.Plotly,\n",
    "    labs = None,\n",
    "    jitter_frac=0.02,\n",
    "    random_seed=42,\n",
    "    **kwargs,\n",
    "    ):\n",
    "\n",
    "    # setup data\n",
    "    data = data.copy()\n",
    "    \n",
    "    # apply jitter if desired - helps spread out points when tightly clustered\n",
    "    if jitter_frac:\n",
    "        np.random.seed(random_seed)\n",
    "        data[x] = jitter(data[x], jitter_frac)\n",
    "        data[y] = jitter(data[y], jitter_frac)\n",
    "            \n",
    "    # convenience variables\n",
    "    data['country_location'] = data['country'] + ' - ' + data['location']\n",
    "    data['size'] = 1  # hack to allow us to control marker size\n",
    "    \n",
    "    # setup plotting options\n",
    "    plot_kwargs = dict(\n",
    "        width=700,\n",
    "        height=500,\n",
    "        template='simple_white',\n",
    "        hover_name='sample_id',\n",
    "        hover_data=[\n",
    "            'sample_id',\n",
    "            'location', \n",
    "            'country', \n",
    "        ],\n",
    "        size='size',\n",
    "        size_max=8,\n",
    "        opacity=0.9,\n",
    "        color_discrete_map = palette,\n",
    "        render_mode='svg'\n",
    "    )\n",
    "    # apply any user overrides\n",
    "    plot_kwargs.update(kwargs)\n",
    "\n",
    "    # 2D scatter plot\n",
    "    fig = px.scatter(data, x=x, y=y, color=colour, labels=labs, **plot_kwargs)\n",
    "    fig.show(config=config)\n",
    "\n",
    "\n",
    "def hash_params(*args, **kwargs):\n",
    "    \"\"\"Helper function to hash analysis parameters.\"\"\"\n",
    "    o = {\n",
    "        'args': args,\n",
    "        'kwargs': kwargs\n",
    "    }\n",
    "    s = json.dumps(o, sort_keys=True).encode()\n",
    "    h = hashlib.md5(s).hexdigest()\n",
    "    return h\n",
    "\n",
    "def load_geno_ds(chrom, \n",
    "                 df_samples,\n",
    "                sample_query=None, \n",
    "                numgenos=None, \n",
    "                sample_list=None, \n",
    "                start=None, \n",
    "                end=None, \n",
    "                min_minor_ac=None,\n",
    "                acconly=True\n",
    "                ):\n",
    "                 # load sample metadata\n",
    "     # Load ds\n",
    "     ds = sg.load_dataset(f'/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/combined_cohorts.{chrom}.zarr')\n",
    "\n",
    "    # Locate selected samples if sample query\n",
    "\n",
    "     if sample_query:\n",
    "        loc_samples = df_samples.query(sample_query).index\n",
    "        df_samples = df_samples.loc[loc_samples, :]\n",
    "        ds = ds.isel(samples=loc_samples)\n",
    "     elif sample_list:\n",
    "          loc_samples = df_samples['sample_id'].isin(sample_list)\n",
    "          df_samples = df_samples.loc[loc_samples, :]\n",
    "          ds = ds.isel(samples=loc_samples)\n",
    "     else:\n",
    "          pass\n",
    "          \n",
    "     \n",
    "     # If minmaf is specified, filter on minmaf  \n",
    "     if min_minor_ac:   \n",
    "        print(f'subsetting to segregating sites')\n",
    "        ac = allel.GenotypeArray(ds['call_genotype']).count_alleles()\n",
    "        macbool = ac[:,1] >= min_minor_ac\n",
    "        print(f'selected {np.sum(macbool)} sites with a min mac > {min_minor_ac}')\n",
    "        ds_analysis = ds.sel(variants=(macbool))\n",
    "     else:\n",
    "        ds_analysis = ds\n",
    "\n",
    "     # Get accessible only\n",
    "     if acconly:\n",
    "        print('subsetting to accessible sites only')\n",
    "        accmask = ds_analysis['is_accessible'].compute()\n",
    "        ds_analysis = ds_analysis.sel(variants=(accmask))\n",
    "     else:\n",
    "            pass\n",
    "     \n",
    "    #if numgenos is set, subset \n",
    "     if numgenos:\n",
    "          ds_analysis = select_random_genos(ds_analysis)\n",
    "     elif start:\n",
    "        print(f\"subsetting haps to range {chrom}:{start}-{end}\")\n",
    "        ds_analysis = ds_analysis.set_index(variants=(\"variant_contig\", \"variant_position\")).sel(variants=(0, slice(start,end)))\n",
    "     else:\n",
    "            pass\n",
    "\n",
    "\n",
    "\n",
    "     #return completed ds\n",
    "     return(df_samples, ds_analysis)\n",
    "\n",
    "def run_pca( \n",
    "    chrom,\n",
    "    df_samples,\n",
    "    sample_query=None,\n",
    "    site_mask=\"gamb_colu_arab\",\n",
    "    min_minor_ac=3,\n",
    "    max_an_missing=0,\n",
    "    n_snps=100000,\n",
    "    analysis_name = 'pca',\n",
    "    results_dir=None,\n",
    "    snp_offset=0,\n",
    "    n_components=10,\n",
    "    sample_list=None):\n",
    "    \"\"\"Main function to run a PCA.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    region : str\n",
    "        Chromosome arm, e.g., '3L', or region, e.g., '3L:1000000-37000000'.\n",
    "    sample_sets : str or list of str, optional\n",
    "        Sample sets to analyse.\n",
    "    sample_query : str, optional\n",
    "        A pandas query string to select specific samples.\n",
    "    site_mask : {'gamb_colu_arab', 'gamb_colu', 'arab'}\n",
    "        Which site mask to apply.\n",
    "    min_minor_ac : int\n",
    "        Minimum minor allele count.\n",
    "    max_an_missing : int\n",
    "        Maximum number of missing allele calls.\n",
    "    n_snps : int\n",
    "        Approximate number of SNPs to use.\n",
    "    snp_offset : int\n",
    "        Offset when thinning SNPs.\n",
    "    n_components : int\n",
    "        Number of PCA components to retain.\n",
    "    sample_list : int\n",
    "       List of samples you want to retain.\n",
    "\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    data : pandas DataFrame\n",
    "        Data frame with one row per sample, including columns \"PC1\", \"PC2\", etc.\n",
    "    evr : numpy array\n",
    "        Explained variance ratio per principal component.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # construct a key to save the results under\n",
    "    results_key = hash_params(\n",
    "        #region=region,\n",
    "        #sample_sets=sample_sets,\n",
    "        analysis_name=analysis_name,\n",
    "        sample_query=sample_query,\n",
    "        site_mask=site_mask,\n",
    "        min_minor_ac=min_minor_ac,\n",
    "        max_an_missing=max_an_missing,\n",
    "        n_snps=n_snps,\n",
    "        snp_offset=snp_offset,\n",
    "        n_components=n_components,\n",
    "        sample_list=sample_list\n",
    "    )\n",
    "\n",
    "    # define paths for results files\n",
    "    data_path = f'{results_dir}/{results_key}-data.csv'\n",
    "    evr_path = f'{results_dir}/{results_key}-evr.npy'\n",
    "\n",
    "    try:\n",
    "        # try to load previously generated results\n",
    "        data = pd.read_csv(data_path)\n",
    "        evr = np.load(evr_path)\n",
    "        return data, evr\n",
    "    except FileNotFoundError:\n",
    "        # no previous results available, need to run analysis\n",
    "        print(f'running analysis: {results_key}')\n",
    "    \n",
    "    print('setting up inputs')\n",
    "\n",
    "    # load sample metadata\n",
    "    \n",
    "    # access SNP calls\n",
    "    sample_df, ds = load_geno_ds(\n",
    "          chrom=chrom,\n",
    "          df_samples=df_samples,\n",
    "          numgenos=n_snps,\n",
    "          min_minor_ac=min_minor_ac,\n",
    "          sample_query=sample_query,\n",
    "          sample_list=sample_list,\n",
    "          acconly = True,\n",
    "    )\n",
    "   \n",
    "\n",
    "    # access SNP genotypes\n",
    "    gt = ds[\"call_genotype\"].data    \n",
    "\n",
    "    print('locating segregating sites within desired frequency range')\n",
    "\n",
    "    # convert to genotype alt counts\n",
    "    gn_seg = allel.GenotypeDaskArray(gt).to_n_alt().compute()\n",
    "    \n",
    "    # remove any edge-case variants where all genotypes are identical\n",
    "    loc_var = np.any(gn_seg != gn_seg[:, 0, np.newaxis], axis=1)\n",
    "    gn_var = np.compress(loc_var, gn_seg, axis=0)\n",
    "\n",
    "    print('running PCA')\n",
    "\n",
    "    # run the PCA\n",
    "    coords, model = allel.pca(gn_var, n_components=n_components)\n",
    "    \n",
    "    #ensure df samplesz are aligned with genos\n",
    "    samples = ds.sample_id.values\n",
    "    sample_df = sample_df.set_index(\"sample_id\").loc[samples].reset_index()\n",
    "\n",
    "    # add PCs to dataframe\n",
    "    # Combine coords and sample metadata.\n",
    "    df_coords = pd.DataFrame(\n",
    "        {f\"PC{i + 1}\": coords[:, i] for i in range(coords.shape[1])}\n",
    "    )\n",
    "    df_pca = sample_df.join(df_coords, how=\"inner\")\n",
    "    \n",
    "    # save results to hash and csv\n",
    "    evr = model.explained_variance_ratio_\n",
    "    df_pca.to_csv(data_path, index=False)\n",
    "    np.save(evr_path, evr)\n",
    "    print(f'saved results: {results_key}')\n",
    "    \n",
    "    return df_pca, evr\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a87362e-8362-4ab8-8f30-38849c008374",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#load data\n",
    "\n",
    "# Load metadata\n",
    "df_samples = pd.read_csv('/Users/dennistpw/Projects/AsGARD/metadata/cease_combinedmetadata_noqc.20250212.csv')\n",
    "\n",
    "# Define results dir\n",
    "outdir = '/Users/dennistpw/Projects/AsGARD/data/results_cache'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65334bcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "model2, var2 = run_pca(chrom='CM023248',\n",
    "        n_snps=100_000,\n",
    "        results_dir=outdir,\n",
    "        min_minor_ac=2,\n",
    "        df_samples=df_samples,\n",
    "        analysis_name='iran1_chr2_next')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bc6e629",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_coords(model2, 'PC1', 'PC2', \n",
    "            colour = 'pop_code', \n",
    "            palette = pop_code_cols,\n",
    "            labs = {'PC1' : f'PC1 ({var2[0]:.2f})',\n",
    "                    'PC2' : f'PC2 ({var2[1]:.2f})',\n",
    "                    'country' : 'Country'},)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b30ed32",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot lower pcs to take a look\n",
    "plot_coords(model2, 'PC3', 'PC4', \n",
    "            colour = 'pop_code', \n",
    "            palette = pop_code_cols,\n",
    "            labs = {'PC3' : f'PC3 ({var2[2]:.2f})',\n",
    "                    'PC4' : f'PC4 ({var2[3]:.2f})',\n",
    "                    'country' : 'Country'},)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52ded87e",
   "metadata": {},
   "source": [
    "The separation between KSA and non-KSA, explaining the most variation in the data, is the most striking. Interestingly, Sudan groups away from most of the other invasive range samples. Sudan is also the most sequenced population in our data, and uneven sampling can lead to distorted PCA results (McVean 2008). What if we randomly downsample our samples?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37b5e059",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_samples.groupby('country').count()['sample_id']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8ac2709",
   "metadata": {},
   "source": [
    "Let's try downsampling to no more than 30 per country"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06680c71",
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped = df_samples.groupby('country')\n",
    "\n",
    "# Apply the sample method to each group\n",
    "sampled = grouped.apply(lambda x: x.sample(min(len(x), 30)))\n",
    "\n",
    "# Remove the group keys\n",
    "sampled = sampled.reset_index(drop=True)\n",
    "\n",
    "inc_sample_list = sampled['sample_id'].tolist()\n",
    "inc_sample_list = [item for item in inc_sample_list if item not in ['DAsgPak010', 'DAsgCPak034', 'Plate2_C2']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af644d87",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_sub_chr2, var_sub_chr2 = run_pca(chrom='CM023248',df_samples = df_samples, results_dir=outdir, analysis_name = '527592', sample_list=inc_sample_list, n_snps=100_000, min_minor_ac=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ded1f11",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_coords(model_sub_chr2, 'PC1', 'PC2', \n",
    "            colour = 'pop_code', \n",
    "            palette = pop_code_cols,\n",
    "            labs = {'PC1' : f'PC1 ({var_sub_chr2[0]*100:.2f}%)',\n",
    "                    'PC2' : f'PC2 ({var_sub_chr2[1]*100:.2f}%)',\n",
    "                    'country' : 'Country',\n",
    "                    'analysis_pop' : 'Population'},)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a29e1ebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_coords(model_sub_chr2, 'PC3', 'PC4', \n",
    "            colour = 'pop_code', \n",
    "            palette = pop_code_cols,\n",
    "            labs = {'PC3' : f'PC3 ({var_sub_chr2[2]:.2f})',\n",
    "                    'PC4' : f'PC4 ({var_sub_chr2[3]:.2f})',\n",
    "                    'country' : 'Country'},)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c7a707b",
   "metadata": {},
   "source": [
    "This is different, Sudan is grouping much more closely to the rest of the invasive samples. Let's examine the other two chromosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb6176a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_sub_chr3, var_sub_chr3 = run_pca(chrom='CM023249',df_samples = df_samples, results_dir=outdir, analysis_name = '396739', sample_list=inc_sample_list, n_snps=100_000, min_minor_ac=2)\n",
    "plot_coords(model_sub_chr3, 'PC1', 'PC2', \n",
    "            colour = 'pop_code', \n",
    "            palette = pop_code_cols,\n",
    "            labs = {'PC1' : f'PC1 ({var_sub_chr3[0]*100:.2f}%)',\n",
    "                    'PC2' : f'PC2 ({var_sub_chr3[1]*100:.2f}%)',\n",
    "                    'country' : 'Country',\n",
    "                    'analysis_pop' : 'Population'},)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "649d0dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_sub_chrX, var_sub_chrX = run_pca(chrom='CM023250',df_samples = df_samples, results_dir=outdir, analysis_name = '0183173', sample_list=inc_sample_list, n_snps=100_000, min_minor_ac=2)\n",
    "plot_coords(model_sub_chrX, 'PC1', 'PC2', \n",
    "            colour = 'pop_code', \n",
    "            palette = pop_code_cols,\n",
    "            labs = {'PC1' : f'PC1 ({var_sub_chrX[0]*100:.2f}%)',\n",
    "                    'PC2' : f'PC2 ({var_sub_chrX[1]*100:.2f}%)',\n",
    "                    'country' : 'Country',\n",
    "                    'analysis_pop' : 'Population'},)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "222abdf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.ticker import LinearLocator, MaxNLocator, FuncFormatter\n",
    "\n",
    "def plot_pcfig(mod, v,  figname, pal='Paired', colour = 'pop_code'):\n",
    "    # Initialize the matplotlib figure\n",
    "    #ig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 6))\n",
    "    sns.set_theme(rc={'figure.figsize':(6,6)}, style=\"ticks\")\n",
    "\n",
    "\n",
    "    scatter_kws = {\n",
    "        's': 150,  # Point size\n",
    "        'edgecolor': 'white',  # Thin white boundary\n",
    "        'linewidth': 0.5,  # Thickness of the boundary\n",
    "        'alpha': 0.8  # Point opacity (80% opacity)\n",
    "    }\n",
    "\n",
    "    ax = plt.gca()\n",
    "\n",
    "    # Plot the first scatterplot\n",
    "    sns.scatterplot(data=mod, x='PC1', y='PC2', hue=colour, ax=ax, palette=pal, **scatter_kws)\n",
    "\n",
    "    # Plot the second scatterplot\n",
    "\n",
    "\n",
    "    # Set titles\n",
    "\n",
    "    # Set axis labels with '%'\n",
    "    ax.set_xlabel(f'PC1 ({(v[0])*100:.2f}%)', fontsize=16)\n",
    "    ax.set_ylabel(f'PC2 ({(v[1])*100:.2f}%)', fontsize=16)\n",
    "\n",
    "    ax.tick_params(labelbottom=False, labelleft=False)    \n",
    "    ax.xaxis.set_major_locator(MaxNLocator(nbins=4, integer=True))  # 5 ticks on the x-axis\n",
    "    ax.yaxis.set_major_locator(MaxNLocator(4))  # 5 ticks on the x-axis\n",
    "    # Remove the legends from individual plots\n",
    "    ax.legend_.remove()\n",
    "# Ensure spines and labels are visible\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    #ax.spines['bottom'].set_visible(True)\n",
    "    #ax.spines['left'].set_visible(True)\n",
    "    ax.spines['left'].set_position(('outward', 10))  # Move the left spine further to the left\n",
    "    ax.spines['bottom'].set_position(('outward', 10))  # Move the left spine further to the left\n",
    "\n",
    "        # Customize tick marks\n",
    "#    ax.tick_params(axis='both', which='both', length=5, labelleft=False)\n",
    "\n",
    "\n",
    "    # Create a single legend\n",
    "    #handles, labels = scatter1.get_legend_handles_labels()\n",
    "    #fig.legend(handles, labels, loc='center right', title='Pop Code', bbox_to_anchor=(1.1, 0.5), borderaxespad=0.)\n",
    "\n",
    "    # Adjust the layout\n",
    "    plt.tight_layout()\n",
    "\n",
    "    plt.savefig(f'../figures/{figname}.svg', format='svg')\n",
    "\n",
    "    # Display the plot\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3e20858",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")  # Options: white, dark, whitegrid, darkgrid, ticks\n",
    "\n",
    "plot_pcfig(model_sub_chr2, var_sub_chr2, pal=pop_code_cols, figname='pca_ab')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf8b71b4",
   "metadata": {},
   "source": [
    "Now let's analyse the invasive range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ee3476d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chr 2 invasive range only\n",
    "mod_invasiverange, var_invasiverange = run_pca(chrom='CM023248',df_samples= df_samples, results_dir=outdir, analysis_name = '2597582', sample_query = 'country == \"Sudan\" | country == \"Ethiopia\" | country == \"Djibouti\" | country == \"Yemen\"', n_snps=100_000, min_minor_ac=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d173824",
   "metadata": {},
   "outputs": [],
   "source": [
    "pinv = plot_pcfig(mod_invasiverange, var_invasiverange, pal=pop_code_cols, figname = 'pca_invasive')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d147b85",
   "metadata": {},
   "source": [
    "Let's replot the chr3 and chrX PCAs for supp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f41211fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_pcfig(model_sub_chr3, var_sub_chr3, pal=pop_code_cols, figname='pca_chr3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3831efa",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_pcfig(model_sub_chrX, var_sub_chrX, pal=pop_code_cols, figname='pca_chrX')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "582b57ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now run PCA for invasive range only on chr 3 and x\n",
    "\n",
    "mod_invasiverange3, var_invasiverange3 = run_pca(chrom='CM023249',df_samples= df_samples, results_dir=outdir, analysis_name = '5757454', sample_query = 'country == \"Sudan\" | country == \"Ethiopia\" | country == \"Djibouti\" | country == \"Yemen\"', n_snps=100_000, min_minor_ac=2)\n",
    "mod_invasiverangeX, var_invasiverangeX = run_pca(chrom='CM023250',df_samples= df_samples, results_dir=outdir, analysis_name = '7587575', sample_query = 'country == \"Sudan\" | country == \"Ethiopia\" | country == \"Djibouti\" | country == \"Yemen\"', n_snps=100_000, min_minor_ac=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07eb5254",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_pcfig(mod_invasiverange3, var_invasiverange3, pal=pop_code_cols, figname='inv_chr3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a83f346c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_pcfig(mod_invasiverangeX, var_invasiverangeX, pal=pop_code_cols, figname='inv_chrX')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cease_local",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
