{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sgkit\n",
    "import allel\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import zarr\n",
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_samples = pd.read_csv('/Users/dennistpw/Projects/AsGARD/metadata/cease_combinedmetadata_noqc.20250212.csv')\n",
    "\n",
    "#palettes\n",
    "pop_code_cols = {\n",
    "    'SAE' : '#6a3d9a', #dark purple\n",
    "    'SAR' : '#cab2d6', #ligher purple\n",
    "    'IRH' : '#c57fc9', #horrible pink\n",
    "    'IRS' : '#c27a88', #salmon\n",
    "    'APA' : '#ff7f00', #orange\n",
    "    'INB' : '#96172e', #darkred\n",
    "    'INM' : '#f03e5e', #lightred\n",
    "    'DJI' : '#507d2a', #sap green\n",
    "    'ETW' : '#a6cee3',#cerulean\n",
    "    'ETB' : '#007272', #cobalt turq\n",
    "    'ETS' : '#33a02c',#green\n",
    "    'SUD' : '#fccf86',#ochre\n",
    "    'YEM' : '#CC7722'#pinkish\n",
    "}\n",
    "\n",
    "# Initialize an empty dictionary to store row indices for each level\n",
    "pop_dict = {}\n",
    "\n",
    "# Iterate through unique levels in the 'factor_column'\n",
    "for level in pop_code_cols.keys():\n",
    "    # Get the row indices where the 'factor_column' matches the current level\n",
    "    indices = df_samples.index[df_samples['pop_code'] == level].tolist()\n",
    "    \n",
    "    # Store the indices in the dictionary with the level as the key\n",
    "    pop_dict[level] = indices\n",
    "\n",
    "# Remove dictionary entries with fewer than 5 values\n",
    "pop_dict = {key: value for key, value in pop_dict.items() if len(value) >= 5}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "annzarr = zarr.open('/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/annotations.zarr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getvariantsbyrange(chrom, start, end, feature):\n",
    "    pos = allel.SortedIndex(annzarr[f'{chrom}/variants/POS'])\n",
    "    loc_region = pos.locate_range(start, end)\n",
    "    range = pos[loc_region]\n",
    "    gt_region = allel.GenotypeArray(annzarr[f'{chrom}/calldata/GT'][loc_region])\n",
    "    ft_region = pd.DataFrame(annzarr[f'{chrom}/variants/{feature}'][loc_region])\n",
    "    ft_id = pd.DataFrame(annzarr[f'{chrom}/variants/ANN_Gene_ID'][loc_region])\n",
    "    im_region = pd.DataFrame(annzarr[f'{chrom}/variants/ANN_Annotation'][loc_region])\n",
    "    annreg = pd.concat([ft_region, im_region, ft_id], axis=1)\n",
    "    annreg['pos'] = range.values\n",
    "    annreg['chrom'] = chrom\n",
    "    acframe = pd.DataFrame(gt_region.count_alleles(), columns=['major', 'minor'])\n",
    "    annreg.columns = ['effect', 'type','id','pos', 'chrom']\n",
    "    annreg = pd.concat([annreg, acframe], axis=1)\n",
    "    return(annreg, gt_region)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define Genes for Heatmap\n",
    "genes_for_heatmap = {\n",
    "    'Ace1' : {'chrom' : 'CM023248', 'start' : 60904014, 'end': 60973023},\n",
    "    'Vgsc' : {'chrom' : 'CM023249', 'start' : 42804885, 'end': 42848176},\n",
    "    'Rdl' : {'chrom' : 'CM023249', 'start' : 8345440, 'end': 8411439},\n",
    "    'Cyp9k1' : {'chrom' : 'CM023250', 'start' : 9713374, 'end': 9729212},\n",
    "    'Cyp6' : {'chrom' : 'CM023248', 'start' : 67470071, 'end': 67514071},\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_plotting_df(chrom, start, end):\n",
    "    x = getvariantsbyrange(chrom,start,end, feature = 'ANN_HGVS_p')\n",
    "    #get subpop dict\n",
    "    maflist = []\n",
    "    subpop_counts = x[1].count_alleles_subpops(pop_dict)\n",
    "    for key in subpop_counts:\n",
    "        maf = pd.DataFrame(subpop_counts[key][:,1] / (subpop_counts[key][:,1] + subpop_counts[key][:,0]))\n",
    "        maflist.append(maf)   \n",
    "    mafdf = pd.concat(maflist, axis=1)\n",
    "    colnames = list(subpop_counts.keys())\n",
    "    mafdf.columns = colnames\n",
    "    mafdf = pd.concat([x[0], mafdf], axis=1)\n",
    "    gene_df = mafdf[mafdf['type'] == 'missense_variant'].round(2)\n",
    "    gene_df = gene_df.drop(['id','pos','chrom','major','minor','type'], axis=1)\n",
    "    gene_df.set_index('effect',inplace=True)\n",
    "    gene_df = gene_df[(gene_df > 0.05).any(axis=1)]\n",
    "    gene_df = gene_df[['SAE','SAR','IRH','IRS','APA','INM','INB','DJI','ETW','ETB','ETS','SUD','YEM']]\n",
    "    return(gene_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "dflist = []\n",
    "for gene in genes_for_heatmap.keys():\n",
    "    df =  return_plotting_df(chrom=genes_for_heatmap[gene]['chrom'],\n",
    "                             start=genes_for_heatmap[gene]['start'],\n",
    "                             end=genes_for_heatmap[gene]['end']\n",
    "                            )\n",
    "    df['gene'] = gene\n",
    "    dflist.append(df)\n",
    "irdf = pd.concat(dflist)\n",
    "irdf.to_csv('../data/irgenes_snps.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now let's take a closer look at metabolics\n",
    "metabolics = {'ASTEI20_043445':'CM023248',\n",
    "'ASTEI20_031508':'CM023249',\n",
    "'ASTEI20_035158':'CM023249',\n",
    "'ASTEI20_040637':'CM023249',\n",
    "'ASTEI20_041705':'CM023249',\n",
    "'ASTEI20_043207':'CM023249',\n",
    "'ASTEI20_044800':'CM023249',\n",
    "'ASTEI20_031989':'CM023249',\n",
    "'ASTEI20_033264':'CM023249',\n",
    "'ASTEI20_035228':'CM023249',\n",
    "'ASTEI20_041777':'CM023249',\n",
    "'ASTEI20_032647':'CM023248',\n",
    "'ASTEI20_035171':'CM023248',\n",
    "'ASTEI20_035508':'CM023248',\n",
    "'ASTEI20_035875':'CM023248',\n",
    "'ASTEI20_036300':'CM023248',\n",
    "'ASTEI20_036499':'CM023248',\n",
    "'ASTEI20_038121':'CM023248',\n",
    "'ASTEI20_040013':'CM023248',\n",
    "'ASTEI20_041713':'CM023248',\n",
    "'ASTEI20_043191':'CM023248',\n",
    "'ASTEI20_031614':'CM023250',\n",
    "'ASTEI20_032673':'CM023249',\n",
    "'ASTEI20_036053':'CM023249',\n",
    "'ASTEI20_037434':'CM023249',\n",
    "'ASTEI20_040345':'CM023249',\n",
    "'ASTEI20_044864':'CM023249',\n",
    "'ASTEI20_044946':'CM023249',\n",
    "'ASTEI20_045507':'CM023249'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene id and chrom\n",
    "gene_id = 'ASTEI20_043445'\n",
    "\n",
    "\n",
    "def get_missense_snps(gene_id, \n",
    "                      chrom,\n",
    "                      min_maf = 0.05\n",
    "                      ):\n",
    "\n",
    "    chrom = metabolics[gene_id]\n",
    "\n",
    "    # Create bool array for filtering on gene id\n",
    "    id_mask = annzarr[f'{chrom}/variants/ANN_Gene_ID'][:] == gene_id\n",
    "\n",
    "    # Get annotation type array\n",
    "    ann = pd.DataFrame(annzarr[f'{chrom}/variants/ANN_Annotation'][:][id_mask])\n",
    "\n",
    "    # Get features\n",
    "    eff = pd.DataFrame(annzarr[f'{chrom}/variants/ANN_HGVS_p'][:][id_mask])\n",
    "\n",
    "    # Get range\n",
    "    pos = pd.DataFrame(annzarr[f'{chrom}/variants/POS'][:][id_mask])\n",
    "\n",
    "    #bind into df\n",
    "    eff_df = pd.concat([pos, eff, ann], axis=1)\n",
    "    eff_df['gene_id'] = gene_id\n",
    "    eff_df['chrom'] = chrom\n",
    "\n",
    "    # Rename cols and reorder\n",
    "    eff_df.columns = ['pos','effect','annotation','gene_id','chrom']\n",
    "    eff_df = eff_df[['chrom','pos','gene_id','annotation','effect']]\n",
    "\n",
    "    # Get genotypes\n",
    "    gt_region = allel.GenotypeArray(annzarr[f'{chrom}/calldata/GT']).compress(id_mask)\n",
    "\n",
    "    # Get per subpop allel counts\n",
    "    subpop_counts = gt_region.count_alleles_subpops(pop_dict)\n",
    "\n",
    "    #convert subpop counts to allele frequencies\n",
    "    maflist = []\n",
    "    for key in subpop_counts:\n",
    "        maf = pd.DataFrame(subpop_counts[key][:,1] / (subpop_counts[key][:,1] + subpop_counts[key][:,0]))\n",
    "        maflist.append(maf) \n",
    "\n",
    "    # Add colnames \n",
    "    maf_df = pd.concat(maflist, axis=1)\n",
    "    maf_df.columns = pop_code_cols.keys()\n",
    "\n",
    "    # Bind all\n",
    "    maf_df_all = pd.concat([eff_df] + [maf_df], axis=1)\n",
    "\n",
    "    # Apply minmaf filter\n",
    "    maf_df_all = maf_df_all[(maf_df_all.iloc[:, 5:] > min_maf).any(axis=1)]\n",
    "\n",
    "    maf_missense = maf_df_all[maf_df_all['annotation'] == 'missense_variant']\n",
    "\n",
    "    return(maf_missense)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = get_missense_snps('ASTEI20_036300','CM023248')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolic_resistance_df_list = []\n",
    "for gene in metabolics.keys():\n",
    "    df = get_missense_snps(gene_id=gene,\n",
    "                           chrom=metabolics[gene]\n",
    "                           )\n",
    "    metabolic_resistance_df_list.append(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolic_ir_df = pd.concat(metabolic_resistance_df_list)\n",
    "metabolic_ir_df.to_csv('~/Projects/AsGARD/data/metabolic_snps.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cease_local",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
