{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Investigating signals of selection in the _An. stephensi_ genome\n",
    "\n",
    "Tristan Dennis, 02/25\n",
    "\n",
    "NOTES:\n",
    "- For the gene annotation, you will want to add your own chrom labels, positions, and genes, and update the both the dict and the code in the plotting loops accordingly\n",
    "\n",
    "TO-DO:\n",
    "- Turn stevegen functions into a library for future projects (budget malariagen for those not lucky enough to have theit taxon in malariagen_data).\n",
    "- Tidy up gene dict.\n",
    "- Turn gene annotation into a function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up environment\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "from matplotlib.ticker import FuncFormatter\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import sgkit as sg\n",
    "import json\n",
    "import numba\n",
    "from scipy.spatial.distance import squareform\n",
    "import scipy.cluster.hierarchy as sch\n",
    "import hashlib\n",
    "import dask\n",
    "import allel; print('scikit-allel', allel.__version__)\n",
    "# quieten dask warnings about large chunks\n",
    "dask.config.set(**{'array.slicing.split_large_chunks': True})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up helper variables and dicts\n",
    "\n",
    "#chrom dict\n",
    "scaflens={'CM023248' : 93706023,\n",
    "'CM023249' : 88747589,\n",
    "'CM023250' : 22713616}\n",
    "\n",
    "df_samples = pd.read_csv('/Users/dennistpw/Projects/AsGARD/metadata/cease_combinedmetadata_noqc.20250212.csv')\n",
    "\n",
    "# Palette - this maps onto the cohort labels in the dataframe above.\n",
    "pop_code_cols = {\n",
    "    'APA' : '#ff7f00', #orange\n",
    "    'SAE' : '#6a3d9a', #dark purple\n",
    "    'SAR' : '#cab2d6', #ligher purple\n",
    "    'IRS' : '#c27a88', #horrible fuschia pink\n",
    "    'IRH' : '#c57fc9', #salmon\n",
    "    'INB' : '#96172e', #darkred\n",
    "    'INM' : '#f03e5e', #lightred\n",
    "    'DJI' : '#507d2a', #sap green\n",
    "    'ETB' : '#007272', #cobalt turq\n",
    "    'ETS' : '#33a02c',#green\n",
    "    'ETW' : '#a6cee3',#cerulean\n",
    "    'SUD' : '#fccf86',#ochre\n",
    "    'YEM' : '#CC7722'#khaki\n",
    "}\n",
    "\n",
    "contigs_dict = {'CM023248':0,\n",
    "                'CM023249':1,\n",
    "                'CM023250':2}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define helper functions for stevegen0.5k\n",
    "\n",
    "#takes ds, returns gwss of chrom and samples if specified\n",
    "def do_gwss(\n",
    "     chrom,\n",
    "     analysis_name = None,\n",
    "     sample_query = None,\n",
    "     sample_list = None,\n",
    "     numgenos = None, \n",
    "     winsize=100,\n",
    "     results_dir=None,\n",
    "     ):       \n",
    "\n",
    "     # construct a key to save the results under\n",
    "     results_key = hash_params(\n",
    "          chrom=chrom,\n",
    "          sample_list=sample_list,\n",
    "          analysis_name=analysis_name,\n",
    "          sample_query=sample_query,\n",
    "          numgenos=numgenos,\n",
    "          winsize=winsize\n",
    "     )\n",
    "\n",
    "     # define paths for results files\n",
    "     data_path = f'{results_dir}/{results_key}-data.csv'\n",
    "\n",
    "     try:\n",
    "          # try to load previously generated results\n",
    "          data = pd.read_csv(data_path)\n",
    "          return data\n",
    "     except FileNotFoundError:\n",
    "          # no previous results available, need to run analysis\n",
    "          print(f'running analysis: {results_key}')\n",
    "\n",
    "     #get data\n",
    "     print('setting up inputs')\n",
    "     df_samples, ds_analysis = load_hap_ds(chrom=chrom,sample_query=sample_query)\n",
    "     \n",
    "     # load haps and do scan\n",
    "     ht = allel.GenotypeArray(ds_analysis['call_genotype'])\n",
    "     ht = ht.to_haplotypes()\n",
    "     pos = ds_analysis['variant_position'].compute()\n",
    "     hstats = allel.moving_garud_h(ht, size=winsize)\n",
    "     x = allel.moving_statistic(pos, statistic=np.mean, size=winsize)       \n",
    "     print(f\"completed scan for {chrom}, window size {winsize}\")\n",
    "     \n",
    "     #make df\n",
    "     h123df = pd.DataFrame({'chrom' : chrom, 'midpos':x, 'h1':hstats[0],'h12':hstats[1], 'h123':hstats[2],'h1_h2':hstats[3]})\n",
    "\n",
    "     # save results\n",
    "     h123df.to_csv(data_path, index=False)\n",
    "     print(f'saved results: {results_key}')\n",
    "\n",
    "     return(h123df)\n",
    "\n",
    "#hashing function\n",
    "def hash_params(*args, **kwargs):\n",
    "    \"\"\"Helper function to hash analysis parameters.\"\"\"\n",
    "    o = {\n",
    "        'args': args,\n",
    "        'kwargs': kwargs\n",
    "    }\n",
    "    s = json.dumps(o, sort_keys=True).encode()\n",
    "    h = hashlib.md5(s).hexdigest()\n",
    "    return h\n",
    "\n",
    "\n",
    "def select_random_genos(\n",
    "                    ds, \n",
    "                    numgenos):\n",
    "                     #selects given number of genos at random\n",
    "                    keep_indices = np.random.choice(ds.call_genotype.shape[0], numgenos, replace=False)\n",
    "                    keep_indices.sort()\n",
    "                    thinned_callset = ds.isel(variants=~keep_indices)\n",
    "                    return(thinned_callset)\n",
    "\n",
    "def load_hap_ds(chrom, \n",
    "                phased=False,\n",
    "                sample_query=None, \n",
    "                numgenos=None, \n",
    "                sample_list=None, \n",
    "                start=None, \n",
    "                end=None, \n",
    "                min_minor_ac=0,\n",
    "                df_samples=df_samples):\n",
    "                 # load sample metadata\n",
    "     #load ds\n",
    "     if phased:\n",
    "          ds = sg.load_dataset(f'/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/combined_cohorts.phased.{chrom}.zarr')\n",
    "     else:\n",
    "          ds = sg.load_dataset(f'/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/combined_cohorts.{chrom}.zarr')\n",
    "\n",
    "     #load annotation ds\n",
    "     #annzarr = zarr.open('/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/annotations.zarr')     #if sample query or list are specified, subset accordingly\n",
    "\n",
    "     if sample_query:\n",
    "          # locate selected samples\n",
    "          loc_samples = df_samples.eval(sample_query).values\n",
    "          df_samples = df_samples.loc[loc_samples, :]\n",
    "          ds = ds.isel(samples=loc_samples)\n",
    "     elif sample_list:\n",
    "          loc_samples = df_samples['sample_id'].isin(sample_list)\n",
    "          df_samples = df_samples.loc[loc_samples, :]\n",
    "          ds = ds.isel(samples=loc_samples)\n",
    "     else:\n",
    "          pass\n",
    "          \n",
    "     #if numgenos is set, subset \n",
    "     if numgenos:\n",
    "          ds_analysis = select_random_genos(ds, numgenos)\n",
    "     else:\n",
    "          ds_analysis = ds\n",
    "\n",
    "     #if region is set, subset to region\n",
    "     if start:\n",
    "               #subset to region of interest\n",
    "              print(f\"subsetting variants to range {chrom}:{start}-{end}\")\n",
    "              ds_analysis = ds_analysis.set_index(variants=(\"variant_position\")).sel(variants=slice(start,end))\n",
    "              #and for annotations\n",
    "              #pos = allel.SortedIndex(annzarr[f'{chrom}/variants/POS'])\n",
    "              #loc_region = pos.locate_range(start, end)\n",
    "     #         im_region = annzarr[f'{chrom}/variants/ANN_Annotation'][loc_region]\n",
    "     #         ft_region = annzarr[f'{chrom}/variants/ANN_HGVS_p'][loc_region]\n",
    "     else:\n",
    "            pass\n",
    "     \n",
    "     #if minmaf is specified, select minmaf     \n",
    "     print(f'subsetting to segregating sites')\n",
    "     ac = allel.GenotypeArray(ds_analysis['call_genotype']).count_alleles()\n",
    "     macbool = ac[:,1] >= min_minor_ac\n",
    "     print(f'selected {np.sum(macbool)} sites with a min mac > {min_minor_ac}')\n",
    "     ds_analysis = ds_analysis.sel(variants=(macbool))\n",
    "    # ft_maf = ft_region[macbool]\n",
    "    # im_maf = im_region[macbool]\n",
    "     #get accessible only\n",
    "     print('subsetting to accessible sites only')\n",
    "     accmask = ds_analysis['is_accessible'].compute()\n",
    "     ds_analysis = ds_analysis.sel(variants=(accmask))\n",
    "     #ft_acc = ft_maf[accmask]\n",
    "     #im_acc = im_maf[accmask]\n",
    "\n",
    "     #return completed ds\n",
    "     return(df_samples, ds_analysis)#, ft_acc, im_acc)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now run selection scans\n",
    "winsize=1500\n",
    "sels = []\n",
    "for pop_code in df_samples.pop_code.unique():\n",
    "    for chrom in ['CM023248', 'CM023249', 'CM023250']:\n",
    "        selscan = do_gwss(chrom = chrom,\n",
    "                        sample_query = f'pop_code == \"{pop_code}\"',\n",
    "                        winsize=winsize,\n",
    "                        results_dir='/Users/dennistpw/Projects/AsGARD/data/selection_20240923',\n",
    "                        analysis_name=f\"{pop_code}.{chrom}.{winsize}.ud2\"\n",
    "                        )\n",
    "        #selscan = pd.DataFrame({'pop' : [analysis_pop], 'chrom':[chrom]})\n",
    "        selscan['pop'] = pop_code\n",
    "        print(f\"completed selscan for {chrom}, {pop_code}, {winsize}\")\n",
    "        sels.append(selscan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare for plotting\n",
    "\n",
    "# Concat selscan output to a big table\n",
    "selscans_df = pd.concat(sels)\n",
    "\n",
    "#prepare values for plotting\n",
    "max_h1_dict = dict(selscans_df.groupby('pop')['h12'].max())\n",
    "max_values = selscans_df.groupby('chrom')['midpos'].max()\n",
    "total_max = max_values.sum()\n",
    "column_widths = (max_values / total_max).values\n",
    "col_var_levels = max_values.index\n",
    "\n",
    "#gene regions (hacky, amend)\n",
    "cyp6_region = {'CM023248': {'x_min': 67473117, 'x_max': 67501071, 'y_min': 0, 'y_max': 1}}\n",
    "ace1_region = {'CM023248': {'x_min': 60904014, 'x_max': 60973023, 'y_min': 0, 'y_max': 1}}\n",
    "vgsc_region = {'CM023249': {'x_min': 42817709, 'x_max': 42817800, 'y_min': 0, 'y_max': 1}}\n",
    "gste_region = {'CM023249': {'x_min': 70572788, 'x_max': 70584603, 'y_min': 0, 'y_max': 1}}\n",
    "rdl_region = {'CM023249': {'x_min': 8345440, 'x_max': 8348441, 'y_min': 0, 'y_max': 1}}\n",
    "cyp9_region = {'CM023250': {'x_min': 9721225, 'x_max': 9722225, 'y_min': 0, 'y_max': 1}}\n",
    "carboxyl_cluster = {'CM023249':{'x_min' : 18784698, 'x_max' :18801820, 'y_min': 0, 'y_max': 1}}\n",
    "coejh_region = {'CM023249':{'x_min' : 26291697, 'x_max' :26310369, 'y_min': 0, 'y_max': 1}}\n",
    "diagk = {'CM023250': {'x_min': 4578144, 'x_max': 4578144, 'y_min': 0, 'y_max': 1}}\n",
    "\n",
    "#order rows for plotting\n",
    "rowvars_reordered = ['SAE', 'SAR', 'IRH', 'IRS','APA', 'DJI', 'ETB', 'ETS', 'ETW', 'SUD','YEM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now do the plotting - big selscan multiplot\n",
    "\n",
    "# Get the unique row vars (populations - amend to whatever dataframe variable you have to row by)\n",
    "row_var_levels = selscans_df['pop'].unique()\n",
    "\n",
    "# Set up global axis params:\n",
    "# Set y labs to 1 dp\n",
    "def format_func(value, tick_number):\n",
    "    return f\"{value:.1f}\"\n",
    "# Function to format x-axis labels as integers\n",
    "def format_x(value, tick_number):\n",
    "    return f\"{int(value / 1e6)}\"  \n",
    "\n",
    "# Initialize the figure and GridSpec\n",
    "fig = plt.figure(figsize=(sum(column_widths) * 40, len(row_var_levels) * 2))  # Plots half as high\n",
    "gs = GridSpec(len(row_var_levels), len(col_var_levels) + 1, width_ratios=list(column_widths) + [0.1])\n",
    "\n",
    "# Create the subplots\n",
    "for row_idx, row_val in enumerate(rowvars_reordered): # Iterate over populations\n",
    "    for col_idx, col_val in enumerate(col_var_levels): #Iterate over per-pop chromosomes\n",
    "        row_colour = pop_code_cols.get(row_val, 'black')  # Look up colour in dictionary\n",
    "        ax = fig.add_subplot(gs[row_idx, col_idx])\n",
    "        subset = selscans_df[(selscans_df['pop'] == row_val) & (selscans_df['chrom'] == col_val)] # Subset df to contain only data for that chrom and cohort\n",
    "        sns.lineplot(data=subset, x='midpos', y='h12', ax=ax, linewidth=1.5, color=row_colour) # Plot lineplot\n",
    "\n",
    "        #ax.fill_between(x=subset['midpos'], y1=subset['h12'], color=row_colour, alpha=0.1)  #  This is nice if you have inversions or massive sweeps where some colour underneath the line is helpful - commented out for me as was too busy\n",
    "\n",
    "        # Tinker with the axes:\n",
    "        ax.yaxis.set_major_formatter(FuncFormatter(format_func))\n",
    "\n",
    "        # Despine\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "        # Tidy up axis ticks and labels\n",
    "        #ax.set_ylim(0, 1)  #set max on y to 1 -  ensures consistency across all plots - easier if your sweeps are all kind of the same size.\n",
    "        ax.set_ylim(0, max_h1_dict[row_val])  #set max on y to be max h1 for each pop - this emphasises the sweep signal in weakly swept pops/regions but we must note in the legend!\n",
    "        plt.yticks([max_h1_dict[row_val]]) # Plot only the max y value as a tick - remove axis clutter - delete or amend as appropriate\n",
    "        plt.ylabel('') # No Y label - will add myself in inkscape\n",
    "\n",
    "\n",
    "        # Remove y-axis for 2nd and 3rd columns\n",
    "        if col_idx > 0:\n",
    "            ax.set_ylabel('')\n",
    "            ax.yaxis.set_visible(False)\n",
    "\n",
    "        for ax in fig.get_axes():\n",
    "            ax.tick_params(axis='both', labelsize=18)  # Increase font size of both x and y tick labels\n",
    "\n",
    "        # Remove x-axis titles and replace with chromosome numbering\n",
    "        ax.set_xlabel('')\n",
    "        if row_idx == 11:\n",
    "            if col_idx == 0:\n",
    "                ax.set_xlabel(\"2\", fontsize=24, fontweight='bold')\n",
    "            elif col_idx == 1:\n",
    "                ax.set_xlabel(\"3\", fontsize=24, fontweight='bold')\n",
    "            elif col_idx == 2:\n",
    "                ax.set_xlabel(\"X\", fontsize=24, fontweight='bold')    # Add the row title\n",
    "\n",
    "\n",
    "                \n",
    "        # Set x-axis ticks at increments of 2e7\n",
    "        max_x = selscans_df[selscans_df['chrom'] == col_val]['midpos'].max()\n",
    "        ax.set_xticks(range(0, int(max_x) + 1, 20_000_000))\n",
    "        ax.xaxis.set_major_formatter(FuncFormatter(format_x))\n",
    "\n",
    "        # Hide x-axis labels for non-bottom plots\n",
    "        if row_idx < len(rowvars_reordered) - 1:\n",
    "            ax.xaxis.set_tick_params(which='both', bottom=False, top=False, labelbottom=False)\n",
    "\n",
    "        #remove x axes except bottom one\n",
    "        if row_idx < len(rowvars_reordered) - 1:\n",
    "            ax.xaxis.set_tick_params(which='both', bottom=False, top=False, labelbottom=False)  # Hide x-axis ticks and labels\n",
    "\n",
    "        plt.subplots_adjust(top=0.85) \n",
    "\n",
    "        # Add row (population) labels\n",
    "        row_ax = fig.add_subplot(gs[row_idx, -1])\n",
    "        row_ax.text(0.5, 0.5, row_val, va='center', ha='center', fontsize=24, rotation=90, transform=row_ax.transAxes)\n",
    "        row_ax.axis('off')\n",
    "\n",
    "        # Adjust position of facet labels (move them closer to the plot)\n",
    "        if row_idx == 0:  # Adjust position for row labels (facet labels)\n",
    "            row_ax.set_position([row_ax.get_position().x0, row_ax.get_position().y0 + 0.05, row_ax.get_position().width, row_ax.get_position().height])\n",
    "\n",
    "        # Now, annotate plots - theere is prob a better way to do this...in a function?\n",
    "\n",
    "        # Add text annotations and lines for IR genes - needs to be done by chromosome/pos/gene mapping in the dict above\n",
    "        if col_val == 'CM023248' and row_idx == 0:\n",
    "            # Cyp6\n",
    "            ax.fill_betweenx([cyp6_region[col_val]['y_min'], cyp6_region[col_val]['y_max']],\n",
    "                cyp6_region[col_val]['x_min'], cyp6_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(cyp6_region[col_val]['x_min'], 1, \"Cyp6\", fontsize=24, rotation=90)\n",
    "            # Ace1\n",
    "            ax.text(ace1_region[col_val]['x_min'], 1, \"Ace1\", fontsize=24, rotation=90)\n",
    "            ax.fill_betweenx([ace1_region[col_val]['y_min'], ace1_region[col_val]['y_max']],\n",
    "                ace1_region[col_val]['x_min'], ace1_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "\n",
    "        elif col_val == 'CM023249' and row_idx == 0:\n",
    "            #Vgsc\n",
    "            ax.fill_betweenx([vgsc_region[col_val]['y_min'], vgsc_region[col_val]['y_max']],\n",
    "                vgsc_region[col_val]['x_min'], vgsc_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(vgsc_region[col_val]['x_min'], 1, \"Vgsc\", fontsize=24, rotation=90)\n",
    "\n",
    "            #Gste\n",
    "            ax.fill_betweenx([gste_region[col_val]['y_min'], gste_region[col_val]['y_max']],\n",
    "                gste_region[col_val]['x_min'], gste_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(gste_region[col_val]['x_min'], 1, \"Gste\", fontsize=24, rotation=90)\n",
    "\n",
    "            #Rdl\n",
    "            ax.fill_betweenx([rdl_region[col_val]['y_min'], rdl_region[col_val]['y_max']],\n",
    "                rdl_region[col_val]['x_min'], rdl_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(rdl_region[col_val]['x_min'],1, \"Rdl\", fontsize=24, rotation=90)\n",
    "\n",
    "            #Coeae\n",
    "            ax.fill_betweenx([carboxyl_cluster[col_val]['y_min'], carboxyl_cluster[col_val]['y_max']],\n",
    "                            carboxyl_cluster[col_val]['x_min'], carboxyl_cluster[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.5)\n",
    "            ax.text(carboxyl_cluster[col_val]['x_min'], 1, \"Coeae\", fontsize=24, rotation=90)\n",
    "\n",
    "            #CoeJH\n",
    "            ax.fill_betweenx([coejh_region[col_val]['y_min'], coejh_region[col_val]['y_max']],\n",
    "                            coejh_region[col_val]['x_min'], coejh_region[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.5)\n",
    "            ax.text(coejh_region[col_val]['x_min'], 1, \"Coejh\", fontsize=24, rotation=90)\n",
    "\n",
    "        elif col_val == 'CM023250' and row_idx == 0:\n",
    "            #Cyp9k1\n",
    "            ax.fill_betweenx([cyp9_region[col_val]['y_min'], cyp9_region[col_val]['y_max']],\n",
    "                cyp9_region[col_val]['x_min'], cyp9_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(cyp9_region[col_val]['x_min'], 1, \"Cyp9k1\", fontsize=24, rotation=90)\n",
    "\n",
    "            #Diagk\n",
    "            ax.fill_betweenx([diagk[col_val]['y_min'], diagk[col_val]['y_max']],\n",
    "                diagk[col_val]['x_min'], diagk[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(diagk[col_val]['x_min'], 1, \"Diagk\", fontsize=24, rotation=90)\n",
    "\n",
    "\n",
    "plt.subplots_adjust(wspace=0, hspace=0.1)  # Reduce column and row padding\n",
    "\n",
    "fig.savefig('../figures/h12.svg')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "#save as svg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Haplotype clustering\n",
    "Now let's take a look at some haplotype clustering dendrograms to see whether regions apparently under selection are shared between populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define some clustering and plotting functions\n",
    "#get dist\n",
    "@numba.njit(parallel=True)\n",
    "def pdist_abs_hamming(X):\n",
    "    n_obs = X.shape[0]\n",
    "    n_ftr = X.shape[1]\n",
    "    out = np.zeros((n_obs, n_obs), dtype=np.int32)\n",
    "    for i in range(n_obs):\n",
    "        x = X[i]\n",
    "        for j in numba.prange(i + 1, n_obs):\n",
    "            y = X[j]\n",
    "            d = 0\n",
    "            for k in range(n_ftr):\n",
    "                if x[k] != y[k]:\n",
    "                    d += 1\n",
    "            out[i, j] = d\n",
    "            out[j, i] = d\n",
    "    return out\n",
    "\n",
    "def plot_dendrogram(h, ax, df_samples):\n",
    "\n",
    "    #Helper vars I am not fussed about\n",
    "    color_threshold=0\n",
    "    above_threshold_color='k'\n",
    "    linkage_method = 'single'\n",
    "    distance_sort=True\n",
    "    count_sort=True\n",
    "\n",
    "    dist = allel.pairwise_distance(h, 'hamming') * h.shape[0]\n",
    "\n",
    "    # Hierarchical clustering.\n",
    "    Z = sch.linkage(dist, method=linkage_method)\n",
    "\n",
    "        #harmonise metadata with samples\n",
    "    # Align sample metadata with haplotypes.\n",
    "\n",
    "    # Repeat the dataframe so there is one row of metadata for each haplotype.\n",
    "    leaf_data = pd.DataFrame(np.repeat(df_samples.values, 2, axis=0))\n",
    "    leaf_data.columns = df_samples.columns\n",
    "\n",
    "    sns.despine(ax=ax, offset=5, bottom=True, top=False)\n",
    "    # Compute the dendrogram but don't plot it.\n",
    "    dend = sch.dendrogram(\n",
    "        Z,\n",
    "        count_sort=count_sort,\n",
    "        distance_sort=distance_sort,\n",
    "        no_plot=False,\n",
    "        color_threshold=color_threshold,\n",
    "        above_threshold_color=above_threshold_color,\n",
    "    )\n",
    "\n",
    "    ax.get_xaxis().set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    sns.despine(ax=None, fig=None, offset=5, left=True, bottom=True, top=True, right=True)\n",
    "    ax.set_ylim(bottom=-2)\n",
    "    ax.set_ylabel('')\n",
    "    ax.autoscale(axis='x', tight=True)\n",
    "\n",
    "    return Z, dend, leaf_data\n",
    "\n",
    "def return_persample_het(haps):\n",
    "    return([np.nanmean(allel.heterozygosity_observed(haps[:, [i], :])) for i in range(haps.shape[1])])\n",
    "\n",
    "\n",
    "def fig_hap_structure(h, df_samples=df_samples, h_display=None, mutations=None, vspans=[[]], cluster_labels=[], figsize=(8, 3), \n",
    "                      fn=None, dpi=150, height_ratios=(0.6, .1, .1), hap_pops=None, legend=False, title=None):\n",
    "    \n",
    "    # create the figure\n",
    "    fig = plt.figure(figsize=figsize)\n",
    "    \n",
    "    # define subplot layout\n",
    "    gs_nrows = 3\n",
    "    gs_ncols = 1\n",
    "    gs = mpl.gridspec.GridSpec(gs_nrows, gs_ncols, hspace=0.04, wspace=0.04,\n",
    "                               height_ratios=height_ratios)\n",
    "    \n",
    "    # get per-sample mean het\n",
    "    het = return_persample_het(h)\n",
    "\n",
    "    df_samples['sample_het'] = het\n",
    "\n",
    "    # make haps\n",
    "    h = h.to_haplotypes()\n",
    "\n",
    "    # dendrogram\n",
    "    ax_dend = fig.add_subplot(gs[0, 0])\n",
    "    z, r, leafdata = plot_dendrogram(h, ax_dend, df_samples)\n",
    "    ax_dend.set_ylim(bottom=-5)\n",
    "    if legend:\n",
    "        handles = [mpl.patches.Patch(color=pop_code_cols[pop], label=pop) for pop in pop_code_cols.keys()]\n",
    "        ax_dend.legend(handles=handles, loc='upper right', bbox_to_anchor=(1, 1), ncol=3)\n",
    "    ax_dend.set_yticklabels(ax_dend.get_yticks().astype(int))\n",
    "    ax_dend.set(xticklabels=[])\n",
    "    ax_dend.set_title(title,fontdict={'fontsize':18}, loc='right')\n",
    "\n",
    "    ax_dend.xaxis.set_tick_params(length=3, pad=2)\n",
    "    ax_dend.yaxis.set_tick_params(length=3, pad=2)\n",
    "    ax_dend.yaxis.set_major_locator(mpl.ticker.MaxNLocator(nbins = 3))\n",
    "\n",
    "\n",
    "    # population colours\n",
    "    ax_pops = fig.add_subplot(gs[1, 0])\n",
    "    if hap_pops is None:\n",
    "        hap_pops = leafdata.pop_code.values\n",
    "    x = hap_pops.take(r['leaves'])\n",
    "    hap_clrs = [pop_code_cols[p] for p in x]\n",
    "    ax_pops.broken_barh(xranges=[(i, 1) for i in range(h.shape[1])], yrange=(0, 1), color=hap_clrs);\n",
    "    sns.despine(ax=ax_pops, offset=2, left=True, bottom=True, top=True, right=True)\n",
    "    ax_pops.set_xticks([])\n",
    "    ax_pops.set_yticks([])\n",
    "    ax_pops.set_xlim(0, h.shape[1])\n",
    "    ax_pops.yaxis.set_label_position('left')\n",
    "    fig.savefig(f'../figures/{title}.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a figure with 8 subplots (4 rows, 2 columns)\n",
    "fig, axs = plt.subplots(nrows=4, ncols=2, figsize=(12, 16))\n",
    "\n",
    "# Flatten axs for easy iteration\n",
    "#axs = axs.flatten()\n",
    "\n",
    "# Loop through genes and plot dendrograms in subplots\n",
    "for i, (gene, params) in enumerate(genes_main.items()):\n",
    "    haps = load_hap_ds(chrom=params['chrom'], start=params['start'], end=params['end'], min_minor_ac=1)\n",
    "    ht = allel.GenotypeArray(haps[1]['call_genotype'])\n",
    "\n",
    "    # Pass the correct subplot axis to the plotting function\n",
    "    fig_hap_structure(h=ht, df_samples=df_samples, title=gene)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cease_local",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
