{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sgkit as sg\n",
    "import allel\n",
    "import plotly.express as px\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from sklearn.utils.extmath import row_norms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Treemix takes a tab delimited file of allele counts by population - let's define our populations here based on the countries and the pca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_chr2 = sg.load_dataset('/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/combined_cohorts.CM023248.zarr')\n",
    "\n",
    "#load and filter metadata\n",
    "#load and filter metadata\n",
    "df_samples = pd.read_csv('/Users/dennistpw/Projects/AsGARD/metadata/cease_combinedmetadata_noqc.20250212.csv')\n",
    "\n",
    "#extract gts\n",
    "gt = allel.GenotypeArray(ds_chr2.call_genotype)\n",
    "\n",
    "\n",
    "#next, ld prune. this takes a wee while"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define stevegen1000 functions\n",
    "def select_random_genos(\n",
    "                    ds, \n",
    "                    numgenos,):\n",
    "                     #selects given number of genos at random\n",
    "                    keep_no = int(numgenos)\n",
    "                    keep_indices = np.random.choice(ds.call_genotype.shape[0], keep_no, replace=False)\n",
    "                    keep_indices.sort()\n",
    "                    thinned_callset = ds.isel(variants=~keep_indices)\n",
    "                    return(thinned_callset)\n",
    "                    \n",
    "\n",
    "def load_geno_ds(chrom, \n",
    "                sample_query=None, \n",
    "                numgenos=None, \n",
    "                sample_list=None, \n",
    "                start=None, \n",
    "                end=None, \n",
    "                min_minor_ac=0,\n",
    "                df_samples=df_samples):\n",
    "                 # load sample metadata\n",
    "     #load ds\n",
    "     ds = sg.load_dataset(f'/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/combined_cohorts.{chrom}.zarr')\n",
    "\n",
    "     #if sample query or list are specified, subset accordingly\n",
    "     if sample_query:\n",
    "          # locate selected samples\n",
    "          loc_samples = df_samples.eval(sample_query).values\n",
    "          df_samples = df_samples.loc[loc_samples, :]\n",
    "          ds = ds.isel(samples=loc_samples)\n",
    "     elif sample_list:\n",
    "          loc_samples = df_samples['sample_id'].isin(sample_list)\n",
    "          df_samples = df_samples.loc[loc_samples, :]\n",
    "          ds = ds.isel(samples=loc_samples)\n",
    "     else:\n",
    "          pass\n",
    "          \n",
    "     #if numgenos is set, subset \n",
    "     if numgenos:\n",
    "          ds_analysis = select_random_genos(ds, numgenos=numgenos)\n",
    "     else:\n",
    "          ds_analysis = ds\n",
    "\n",
    "     #if region is set, subset to region\n",
    "     if start:\n",
    "               #subset to region of interest\n",
    "              print(f\"subsetting genos to range {chrom}:{start}-{end}\")\n",
    "              ds_analysis = ds_analysis.set_index(variants=(\"variant_contig\", \"variant_position\")).sel(variants=(0, slice(start,end)))\n",
    "     else:\n",
    "            pass\n",
    "     \n",
    "     #if minmaf is specified, select minmaf     \n",
    "     print(f'subsetting to segregating sites')\n",
    "     ac = allel.GenotypeArray(ds_analysis['call_genotype']).count_alleles()\n",
    "     macbool = ac[:,1] >= min_minor_ac\n",
    "     print(f'selected {np.sum(macbool)} sites with a min mac > {min_minor_ac}')\n",
    "     ds_analysis = ds_analysis.sel(variants=(macbool))\n",
    "\n",
    "     #get accessible only\n",
    "     print('subsetting to accessible sites only')\n",
    "     accmask = ds_analysis['is_accessible'].compute()\n",
    "     ds_analysis = ds_analysis.sel(variants=(accmask))\n",
    "\n",
    "     #return completed ds\n",
    "     return(df_samples, ds_analysis)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subsetting to segregating sites\n",
      "selected 1000000 sites with a min mac > 1\n",
      "subsetting to accessible sites only\n"
     ]
    }
   ],
   "source": [
    "#Get ds and df \n",
    "df, ds = load_geno_ds(\n",
    "             numgenos = 1000000,\n",
    "             min_minor_ac=1,\n",
    "             chrom='CM023248',\n",
    "             df_samples=df_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize an empty dictionary to store row indices for each level\n",
    "pop_dict = {}\n",
    "\n",
    "# Iterate through unique levels in the 'factor_column'\n",
    "for level in df_samples['pop_code'].unique():\n",
    "    # Get the row indices where the 'factor_column' matches the current level\n",
    "    indices = df_samples.index[df_samples['pop_code'] == level].tolist()\n",
    "    \n",
    "    # Store the indices in the dictionary with the level as the key\n",
    "    pop_dict[level] = indices\n",
    "\n",
    "# Remove dictionary entries with fewer than 5 values\n",
    "pop_dict = {key: value for key, value in pop_dict.items() if len(value) >= 5}\n",
    "\n",
    "#get all pops too\n",
    "#pop_dict['all']  = df_samples.index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt = allel.GenotypeArray(ds.call_genotype)\n",
    "ac_subpop = gt.count_alleles_subpops(pop_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count alleles by pop & convert to table\n",
    "ac_subpop = gt.count_alleles_subpops(pop_dict)\n",
    "#get segregating variants only\n",
    "#is_seg = ac_subpop['all'].is_segregating()[:]\n",
    "\n",
    "combined_data = {}\n",
    "\n",
    "# Process each item in the dictionary\n",
    "for name, array in ac_subpop.items():\n",
    "    # Check if the array has at least two columns\n",
    "\n",
    "    #get seg sites\n",
    "   #array = array.compress(is_seg)\n",
    "    if array.shape[1] < 2:\n",
    "        raise ValueError(f\"Array {name} does not have at least two columns\")\n",
    "    \n",
    "    # Combine each element of the two columns into a single string\n",
    "    combined_array = np.array([f\"{row[0]},{row[1]}\" for row in array])\n",
    "    \n",
    "    # Store the resulting array in the combined_data dictionary\n",
    "    combined_data[name] = combined_array\n",
    "\n",
    "# Convert the combined data into a DataFrame\n",
    "ac_df = pd.DataFrame(combined_data)\n",
    "\n",
    "##subsample randomly instead of ld pruning\n",
    "#df_sample = ac_df.sample(frac=0.1)\n",
    "\n",
    "#df_sample = df_sample.drop('all', axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#try to permute across different randomly chosen sets (10 times?)\n",
    "ac_df.to_csv('/Users/dennistpw/Projects/AsGARD/data/TreeMix_20240926/afs_bypop.txt', sep='\\t', quoting=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-09-26 13:43:08--  https://bitbucket.org/nygcresearch/treemix/downloads/treemix-1.13.tar.gz\n",
      "Resolving bitbucket.org (bitbucket.org)... 185.166.142.21, 185.166.142.23, 185.166.142.22, ...\n",
      "Connecting to bitbucket.org (bitbucket.org)|185.166.142.21|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://bbuseruploads.s3.amazonaws.com/915ee537-3309-4916-8689-5a97111dc17c/downloads/bb0e7778-4379-4820-9e81-9ac3a0c926dc/treemix-1.13.tar.gz?response-content-disposition=attachment%3B%20filename%3D%22treemix-1.13.tar.gz%22&response-content-encoding=None&AWSAccessKeyId=ASIA6KOSE3BNALVYUHQT&Signature=5%2B7Kze7D3y4MvHiDZmAh1scGdbw%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEMb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJIMEYCIQCz20AUt%2BfPJ1RgOHtVEV1PEi7M2mODPcyz1whTIg3omwIhANtj5t6qSEQRDpSqRCIFRYG96tQOG%2FAXXzw0cfKv8t4OKrACCP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMOTg0NTI1MTAxMTQ2Igy10wr62f%2B%2BBOJBCooqhALR0MouypDvKBW7jkXae5n6VDb7gSCJSUhbfeI8oRdOHOW0WPcj5a%2FHe4Zz7f6YXElu6gM6hLj2P4gdx%2FF9DFM0a0QAwjy4sm0TesMyQ3linGJxBUgISh%2Fyb7RkcTp%2BjY0zuj2Kz6hg1apR4tQZI7z0CI0knr9zUqCAiRZlAN8wbvT63EYDhsCzHc1oBptlp2qluiiMSmtCnKWGak%2F9uNqVBaT3BlMsq6QTOsPl2Qmy8aj%2ByGdmEssV4ELuvjC5aivE8guHClPqxkwa0eDXUNd9fYOSm2mkET7JhhSNy%2FWRGAb13KeoMUTo3Izjix802caMymGaX1jSFpV9%2BuHpTnhzTYUa1TDa6NO3BjqcAX2SIkXr0da6eTfqOpXpDnNMJknCEgUmZ5QmiUSBX4m33EhiHM7slAWe79fWyzI%2B8pJ5VJo%2F%2FdVdwF0eXoidlXj%2BOBwqC4JAEHWbvpPfCqneQaF7yit9dzCZb%2FlkVzU9o54zi6u3VMy%2FhunAGL4noGvB%2FomL98AsFgVpihJ6QiCpb8PVkuv3e%2FKy2CXwVlmOzYYKUyg9Ukieeq4U4A%3D%3D&Expires=1727331170 [following]\n",
      "--2024-09-26 13:43:09--  https://bbuseruploads.s3.amazonaws.com/915ee537-3309-4916-8689-5a97111dc17c/downloads/bb0e7778-4379-4820-9e81-9ac3a0c926dc/treemix-1.13.tar.gz?response-content-disposition=attachment%3B%20filename%3D%22treemix-1.13.tar.gz%22&response-content-encoding=None&AWSAccessKeyId=ASIA6KOSE3BNALVYUHQT&Signature=5%2B7Kze7D3y4MvHiDZmAh1scGdbw%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEMb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJIMEYCIQCz20AUt%2BfPJ1RgOHtVEV1PEi7M2mODPcyz1whTIg3omwIhANtj5t6qSEQRDpSqRCIFRYG96tQOG%2FAXXzw0cfKv8t4OKrACCP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMOTg0NTI1MTAxMTQ2Igy10wr62f%2B%2BBOJBCooqhALR0MouypDvKBW7jkXae5n6VDb7gSCJSUhbfeI8oRdOHOW0WPcj5a%2FHe4Zz7f6YXElu6gM6hLj2P4gdx%2FF9DFM0a0QAwjy4sm0TesMyQ3linGJxBUgISh%2Fyb7RkcTp%2BjY0zuj2Kz6hg1apR4tQZI7z0CI0knr9zUqCAiRZlAN8wbvT63EYDhsCzHc1oBptlp2qluiiMSmtCnKWGak%2F9uNqVBaT3BlMsq6QTOsPl2Qmy8aj%2ByGdmEssV4ELuvjC5aivE8guHClPqxkwa0eDXUNd9fYOSm2mkET7JhhSNy%2FWRGAb13KeoMUTo3Izjix802caMymGaX1jSFpV9%2BuHpTnhzTYUa1TDa6NO3BjqcAX2SIkXr0da6eTfqOpXpDnNMJknCEgUmZ5QmiUSBX4m33EhiHM7slAWe79fWyzI%2B8pJ5VJo%2F%2FdVdwF0eXoidlXj%2BOBwqC4JAEHWbvpPfCqneQaF7yit9dzCZb%2FlkVzU9o54zi6u3VMy%2FhunAGL4noGvB%2FomL98AsFgVpihJ6QiCpb8PVkuv3e%2FKy2CXwVlmOzYYKUyg9Ukieeq4U4A%3D%3D&Expires=1727331170\n",
      "Resolving bbuseruploads.s3.amazonaws.com (bbuseruploads.s3.amazonaws.com)... 16.182.104.1, 52.216.220.17, 52.216.37.217, ...\n",
      "Connecting to bbuseruploads.s3.amazonaws.com (bbuseruploads.s3.amazonaws.com)|16.182.104.1|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 185363 (181K) [application/x-tar]\n",
      "Saving to: ‘treemix-1.13.tar.gz.1’\n",
      "\n",
      "treemix-1.13.tar.gz  27%[====>               ]  49.50K  35.3KB/s               "
     ]
    }
   ],
   "source": [
    "#Now let's download and install TreeMix\n",
    "!wget https://bitbucket.org/nygcresearch/treemix/downloads/treemix-1.13.tar.gz\n",
    "!tar -xvf treemix-1.13.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#run treemix using the shell scripts\n",
    "#find optimal M if any using R script\n",
    "#optimal M looks like 0/1?\n",
    "#high data robustness...may plot without migration edges"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cease_local",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
