{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "import sgkit as sg\n",
    "import json\n",
    "import numba\n",
    "import scipy\n",
    "from scipy.spatial.distance import squareform\n",
    "from scipy.spatial import distance\n",
    "import scipy.cluster.hierarchy as sch\n",
    "import hashlib\n",
    "import itertools as it\n",
    "import plotly.express as px\n",
    "from IPython.display import Image\n",
    "import xarray\n",
    "import dask\n",
    "sns.set_style('white')\n",
    "sns.set_style('ticks')\n",
    "sns.set_context('notebook')\n",
    "import zarr\n",
    "import allel; print('scikit-allel', allel.__version__)\n",
    "import os\n",
    "from multiprocessing.pool import ThreadPool\n",
    "dask.config.set(pool=ThreadPool(20))\n",
    "from dask.diagnostics import ProgressBar\n",
    "# quieten dask warnings about large chunks\n",
    "dask.config.set(**{'array.slicing.split_large_chunks': True})\n",
    "\n",
    "#load and filter metadata\n",
    "#df_samples[~df_samples['sample_id'].isin(missing_indv)]\n",
    "#metadata = df_samples[~df_samples['sample_id'].isin(missing_indv)]\n",
    "#metadata.reset_index(inplace=True)\n",
    "\n",
    "#chrom dict\n",
    "scaflens={'CM023248' : 93706023,\n",
    "'CM023249' : 88747589,\n",
    "'CM023250' : 22713616}\n",
    "\n",
    "df_samples = pd.read_csv('/Users/dennistpw/Projects/AsGARD/metadata/cease_combinedmetadata_qcpass.20240914.csv')\n",
    "\n",
    "analysis_popcols =  {'saudi_e': '#6a3d9a',\n",
    "'saudi_r': '#cab2d6',\n",
    "'india_m': '#a6cee3',\n",
    " 'india_b': '#1f78b4',\n",
    " 'afgh_pak': '#ff7f00',\n",
    " 'djibouti': '#e31a1c',\n",
    " 'ethiopia_n': '#33a02c',\n",
    " 'ethiopia_so': '#b2df8a',\n",
    " 'yemen': '#fdbf6f',\n",
    " 'sudan': '#fb9a99'}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define stevegen1000 functions\n",
    "def select_random_genos(\n",
    "                    ds, \n",
    "                    numgenos=100_000):\n",
    "                     #selects given number of genos at random\n",
    "                    keep_no = int(100_000)\n",
    "                    keep_indices = np.random.choice(ds.call_genotype.shape[0], 100_000, replace=False)\n",
    "                    keep_indices.sort()\n",
    "                    thinned_callset = ds.isel(variants=~keep_indices)\n",
    "                    return(thinned_callset)\n",
    "\n",
    "def load_hap_ds(chrom, \n",
    "                sample_query=None, \n",
    "                numgenos=None, \n",
    "                sample_list=None, \n",
    "                start=None, \n",
    "                end=None, \n",
    "                min_minor_ac=0,\n",
    "                df_samples=df_samples):\n",
    "                 # load sample metadata\n",
    "     #load ds\n",
    "     ds = sg.load_dataset(f'/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/combined_cohorts.phased.{chrom}.zarr')\n",
    "\n",
    "     #if sample query or list are specified, subset accordingly\n",
    "     if sample_query:\n",
    "          # locate selected samples\n",
    "          loc_samples = df_samples.eval(sample_query).values\n",
    "          df_samples = df_samples.loc[loc_samples, :]\n",
    "          ds = ds.isel(samples=loc_samples)\n",
    "     elif sample_list:\n",
    "          loc_samples = df_samples['sample_id'].isin(sample_list)\n",
    "          df_samples = df_samples.loc[loc_samples, :]\n",
    "          ds = ds.isel(samples=loc_samples)\n",
    "     else:\n",
    "          pass\n",
    "          \n",
    "     #if numgenos is set, subset \n",
    "     if numgenos:\n",
    "          ds_analysis = select_random_genos(ds)\n",
    "     else:\n",
    "          ds_analysis = ds\n",
    "\n",
    "     #if region is set, subset to region\n",
    "     if start:\n",
    "               #subset to region of interest\n",
    "              print(f\"subsetting haps to range {chrom}:{start}-{end}\")\n",
    "              ds_analysis = ds_analysis.set_index(variants=(\"variant_contig\", \"variant_position\")).sel(variants=(0, slice(start,end)))\n",
    "     else:\n",
    "            pass\n",
    "     \n",
    "     #if minmaf is specified, select minmaf     \n",
    "     print(f'subsetting to segregating sites')\n",
    "     ac = allel.GenotypeArray(ds_analysis['call_genotype']).count_alleles()\n",
    "     macbool = ac[:,1] >= min_minor_ac\n",
    "     print(f'selected {np.sum(macbool)} sites with a min mac > {min_minor_ac}')\n",
    "     ds_analysis = ds_analysis.sel(variants=(macbool))\n",
    "\n",
    "     #get accessible only\n",
    "     print('subsetting to accessible sites only')\n",
    "     accmask = ds_analysis['is_accessible'].compute()\n",
    "     ds_analysis = ds_analysis.sel(variants=(accmask))\n",
    "\n",
    "     #return completed ds\n",
    "     return(df_samples, ds_analysis)\n",
    "            \n",
    "#hashing function\n",
    "def hash_params(*args, **kwargs):\n",
    "    \"\"\"Helper function to hash analysis parameters.\"\"\"\n",
    "    o = {\n",
    "        'args': args,\n",
    "        'kwargs': kwargs\n",
    "    }\n",
    "    s = json.dumps(o, sort_keys=True).encode()\n",
    "    h = hashlib.md5(s).hexdigest()\n",
    "    return h\n",
    "\n",
    "#takes ds, returns gwss of chrom and samples if specified\n",
    "def do_gwss(\n",
    "     chrom,\n",
    "     analysis_name = None,\n",
    "     sample_query = None,\n",
    "     sample_list = None,\n",
    "     numgenos = None, \n",
    "     winsize=100,\n",
    "     results_dir=None,\n",
    "     ):       \n",
    "\n",
    "     # construct a key to save the results under\n",
    "     results_key = hash_params(\n",
    "          chrom=chrom,\n",
    "          sample_list=sample_list,\n",
    "          analysis_name=analysis_name,\n",
    "          sample_query=sample_query,\n",
    "          numgenos=numgenos,\n",
    "          winsize=winsize\n",
    "     )\n",
    "\n",
    "     # define paths for results files\n",
    "     data_path = f'{results_dir}/{results_key}-data.csv'\n",
    "\n",
    "     try:\n",
    "          # try to load previously generated results\n",
    "          data = pd.read_csv(data_path)\n",
    "          return data\n",
    "     except FileNotFoundError:\n",
    "          # no previous results available, need to run analysis\n",
    "          print(f'running analysis: {results_key}')\n",
    "\n",
    "     #get data\n",
    "     print('setting up inputs')\n",
    "     df_samples, ds_analysis = load_hap_ds(chrom=chrom,sample_query=sample_query)\n",
    "     \n",
    "     # load haps and do scan\n",
    "     ht = allel.GenotypeArray(ds_analysis['call_genotype'])\n",
    "     ht = ht.to_haplotypes()\n",
    "     pos = ds_analysis['variant_position'].compute()\n",
    "     hstats = allel.moving_garud_h(ht, size=winsize)\n",
    "     x = allel.moving_statistic(pos, statistic=np.mean, size=winsize)       \n",
    "     print(f\"completed scan for {chrom}, window size {winsize}\")\n",
    "     \n",
    "     #make df\n",
    "     h123df = pd.DataFrame({'chrom' : chrom, 'midpos':x, 'h1':hstats[0],'h12':hstats[1], 'h123':hstats[2],'h1_h2':hstats[3]})\n",
    "\n",
    "     # save results\n",
    "     h123df.to_csv(data_path, index=False)\n",
    "     print(f'saved results: {results_key}')\n",
    "\n",
    "     return(h123df)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "winsize=1000\n",
    "sels = []\n",
    "for analysis_pop in df_samples.analysis_pop.unique():\n",
    "    for chrom in ['CM023248', 'CM023249', 'CM023250']:\n",
    "        selscan = do_gwss(chrom = chrom,\n",
    "                        sample_query = f'analysis_pop == \"{analysis_pop}\"',\n",
    "                        winsize=winsize,\n",
    "                        results_dir='/Users/dennistpw/Projects/AsGARD/data/selection_20240923',\n",
    "                        analysis_name=f\"{analysis_pop}.{chrom}.{winsize}.ud2\"\n",
    "                        )\n",
    "        #selscan = pd.DataFrame({'pop' : [analysis_pop], 'chrom':[chrom]})\n",
    "        selscan['pop'] = analysis_pop\n",
    "        print(f\"completed selscan for {chrom}, {analysis_pop}, {winsize}\")\n",
    "        sels.append(selscan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare for plotting\n",
    "\n",
    "#import useful libs\n",
    "from matplotlib.ticker import FuncFormatter\n",
    "from matplotlib.gridspec import GridSpec\n",
    "\n",
    "# Concat to big table\n",
    "selscans_df = pd.concat(sels)\n",
    "\n",
    "#prepare values for plotting\n",
    "max_h1_dict = dict(selscans_df.groupby('pop')['h1'].max())\n",
    "max_values = selscans_df.groupby('chrom')['midpos'].max()\n",
    "total_max = max_values.sum()\n",
    "column_widths = (max_values / total_max).values\n",
    "col_var_levels = max_values.index\n",
    "\n",
    "#gene regions\n",
    "cyp6_region = {'CM023248': {'x_min': 67473117, 'x_max': 67501071, 'y_min': 0, 'y_max': 1}}\n",
    "ace1_region = {'CM023248': {'x_min': 60916071, 'x_max': 60917000, 'y_min': 0, 'y_max': 1}}\n",
    "vgsc_region = {'CM023249': {'x_min': 42817709, 'x_max': 42817800, 'y_min': 0, 'y_max': 1}}\n",
    "gste_region = {'CM023249': {'x_min': 70572788, 'x_max': 70584603, 'y_min': 0, 'y_max': 1}}\n",
    "rdl_region = {'CM023249': {'x_min': 8345440, 'x_max': 8348441, 'y_min': 0, 'y_max': 1}}\n",
    "cyp9_region = {'CM023250': {'x_min': 9721225, 'x_max': 9722225, 'y_min': 0, 'y_max': 1}}\n",
    "carboxyl_cluster = {'CM023249':{'x_min' : 18784698, 'x_max' :18801820, 'y_min': 0, 'y_max': 1}}\n",
    "COEJHE5E_region = {'CM023249':{'x_min' : 26291697, 'x_max' :26310369, 'y_min': 0, 'y_max': 1}}\n",
    "diagk = {'CM023250': {'x_min': 4578144, 'x_max': 4578144, 'y_min': 0, 'y_max': 1}}\n",
    "\n",
    "#order rows for plotting\n",
    "rowvars_reordered = ['saudi_e', 'saudi_r', 'afgh_pak', 'india_b', 'india_m', 'djibouti', 'ethiopia_n', 'ethiopia_so', 'sudan', 'yemen']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the unique row and column variables\n",
    "row_var_levels = selscans_df['pop'].unique()\n",
    "\n",
    "#set y labs to 2 dp\n",
    "def format_func(value, tick_number):\n",
    "    return f\"{value:.2f}\"\n",
    "\n",
    "# Initialize the figure and GridSpec\n",
    "fig = plt.figure(figsize=(sum(column_widths) * 25, len(row_var_levels) * 1.5))  # Plots half as high\n",
    "gs = GridSpec(len(row_var_levels), len(col_var_levels) + 1, width_ratios=list(column_widths) + [0.1])\n",
    "\n",
    "# Create the subplots\n",
    "for row_idx, row_val in enumerate(rowvars_reordered):\n",
    "    for col_idx, col_val in enumerate(col_var_levels):\n",
    "        row_colour = analysis_popcols.get(row_val, 'black')  # Default color if not specified\n",
    "\n",
    "        ax = fig.add_subplot(gs[row_idx, col_idx])\n",
    "        subset = selscans_df[(selscans_df['pop'] == row_val) & (selscans_df['chrom'] == col_val)]\n",
    "        sns.lineplot(data=subset, x='midpos', y='h1', ax=ax, linewidth=0.8, color=row_colour)\n",
    "\n",
    "        ax.fill_between(x=subset['midpos'], y1=subset['h1'], color=row_colour, alpha=0.1)  # Adjust alpha for transparency if needed\n",
    "\n",
    "\n",
    "        #Let's start by tinkering with the axes\n",
    "\n",
    "        #set max on y to be max h1 for each pop - this emphasises the sweep signal but we must note in the legend!\n",
    "        ax.set_ylim(0, max_h1_dict[row_val])\n",
    "        ax.yaxis.set_major_formatter(FuncFormatter(format_func))\n",
    "        #set y max as ticks\n",
    "        ax.set_yticks([0, max_h1_dict[row_val]])\n",
    "        ax.set_ylabel(\"\")\n",
    "        plt.subplots_adjust(top=0.85) \n",
    "\n",
    "        #now we want to remove clutter as much as possible\n",
    "        #despine\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "        ax.spines['bottom'].set_visible(False)\n",
    "\n",
    "        #remove x axes except bottom one\n",
    "        if row_idx < len(rowvars_reordered) - 1:\n",
    "            ax.xaxis.set_tick_params(which='both', bottom=False, top=False, labelbottom=False)  # Hide x-axis ticks and labels\n",
    "\n",
    "        # Remove y-axis for 2nd and 3rd columns\n",
    "        if col_idx > 0:\n",
    "            ax.set_ylabel('')\n",
    "            ax.yaxis.set_visible(False)\n",
    "\n",
    "        # Remove x-axis titles and replace with chromosome numbering\n",
    "        ax.set_xlabel('')\n",
    "        if row_idx == 9:\n",
    "            if col_idx == 0:\n",
    "                ax.set_xlabel(\"2\", fontsize=24, fontweight='bold')\n",
    "            elif col_idx == 1:\n",
    "                ax.set_xlabel(\"3\", fontsize=24, fontweight='bold')\n",
    "            elif col_idx == 2:\n",
    "                ax.set_xlabel(\"X\", fontsize=24, fontweight='bold')    # Add the row title\n",
    "\n",
    "        #add row (population) labels\n",
    "        row_ax = fig.add_subplot(gs[row_idx, -1])\n",
    "        row_ax.text(0.5, 0.5, row_val, va='center', ha='center', fontsize=18, rotation=0, transform=row_ax.transAxes)\n",
    "        row_ax.axis('off')\n",
    "\n",
    "        #now, let's annotate our plots\n",
    "\n",
    "        #add text annotations and lines for IR genes\n",
    "        if col_val == 'CM023248' and row_idx == 0:\n",
    "            # Cyp6\n",
    "            ax.fill_betweenx([cyp6_region[col_val]['y_min'], cyp6_region[col_val]['y_max']],\n",
    "                cyp6_region[col_val]['x_min'], cyp6_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(cyp6_region[col_val]['x_min'], 1, \"Cyp6\", fontsize=18, rotation=90)\n",
    "            # Ace1\n",
    "            ax.text(ace1_region[col_val]['x_min'], 1, \"Ace1\", fontsize=18, rotation=90)\n",
    "            ax.fill_betweenx([ace1_region[col_val]['y_min'], ace1_region[col_val]['y_max']],\n",
    "                ace1_region[col_val]['x_min'], ace1_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "\n",
    "        elif col_val == 'CM023249' and row_idx == 0:\n",
    "            #Vgsc\n",
    "            ax.fill_betweenx([vgsc_region[col_val]['y_min'], vgsc_region[col_val]['y_max']],\n",
    "                vgsc_region[col_val]['x_min'], vgsc_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(vgsc_region[col_val]['x_min'], 1, \"Vgsc\", fontsize=18, rotation=90)\n",
    "\n",
    "            #Gste\n",
    "            ax.fill_betweenx([gste_region[col_val]['y_min'], gste_region[col_val]['y_max']],\n",
    "                gste_region[col_val]['x_min'], gste_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(gste_region[col_val]['x_min'], 1, \"Gste\", fontsize=18, rotation=90)\n",
    "\n",
    "            #Rdl\n",
    "            ax.fill_betweenx([rdl_region[col_val]['y_min'], rdl_region[col_val]['y_max']],\n",
    "                rdl_region[col_val]['x_min'], rdl_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(rdl_region[col_val]['x_min'],1, \"Rdl\", fontsize=18, rotation=90)\n",
    "\n",
    "            #Coeae\n",
    "            ax.fill_betweenx([carboxyl_cluster[col_val]['y_min'], carboxyl_cluster[col_val]['y_max']],\n",
    "                            carboxyl_cluster[col_val]['x_min'], carboxyl_cluster[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.5)\n",
    "            ax.text(carboxyl_cluster[col_val]['x_min'], 1, \"Coeae\", fontsize=18, rotation=90)\n",
    "\n",
    "            #CoeJH\n",
    "            ax.fill_betweenx([COEJHE5E_region[col_val]['y_min'], COEJHE5E_region[col_val]['y_max']],\n",
    "                            COEJHE5E_region[col_val]['x_min'], COEJHE5E_region[col_val]['x_max'],\n",
    "                            color='gray', alpha=0.5)\n",
    "            ax.text(COEJHE5E_region[col_val]['x_min'], 1, \"Coejh\", fontsize=18, rotation=90)\n",
    "\n",
    "        elif col_val == 'CM023250' and row_idx == 0:\n",
    "            #Cyp9k1\n",
    "            ax.fill_betweenx([cyp9_region[col_val]['y_min'], cyp9_region[col_val]['y_max']],\n",
    "                cyp9_region[col_val]['x_min'], cyp9_region[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(cyp9_region[col_val]['x_min'], 1, \"Cyp9k1\", fontsize=18, rotation=90)\n",
    "\n",
    "            #Diagk\n",
    "            ax.fill_betweenx([diagk[col_val]['y_min'], diagk[col_val]['y_max']],\n",
    "                diagk[col_val]['x_min'], diagk[col_val]['x_max'],\n",
    "                color='gray', alpha=0.5)\n",
    "            ax.text(diagk[col_val]['x_min'], 1, \"Diagk\", fontsize=18, rotation=90)\n",
    "\n",
    "#Add master Y labels\n",
    "fig.text(0.08, 0.5, 'H1', va='center', rotation='vertical', fontweight='bold', fontsize=18)\n",
    "\n",
    "plt.subplots_adjust(wspace=0, hspace=0.2)  # Reduce column and row padding\n",
    "\n",
    "fig.savefig('../figures/h12.svg')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "#save as svg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Haplotype clustering\n",
    "Now let's take a look at some haplotype clustering dendrograms to see whether regions apparently under selection are shared between populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define some clustering and plotting functions\n",
    "#get dist\n",
    "@numba.njit(parallel=True)\n",
    "def pdist_abs_hamming(X):\n",
    "    n_obs = X.shape[0]\n",
    "    n_ftr = X.shape[1]\n",
    "    out = np.zeros((n_obs, n_obs), dtype=np.int32)\n",
    "    for i in range(n_obs):\n",
    "        x = X[i]\n",
    "        for j in numba.prange(i + 1, n_obs):\n",
    "            y = X[j]\n",
    "            d = 0\n",
    "            for k in range(n_ftr):\n",
    "                if x[k] != y[k]:\n",
    "                    d += 1\n",
    "            out[i, j] = d\n",
    "            out[j, i] = d\n",
    "    return out\n",
    "\n",
    "def plot_dendrogram(h, ax, df_samples):\n",
    "\n",
    "    #Helper vars I am not fussed about\n",
    "    color_threshold=0\n",
    "    above_threshold_color='k'\n",
    "    linkage_method = 'single'\n",
    "    distance_sort=False\n",
    "    count_sort=False\n",
    "\n",
    "    dist = allel.pairwise_distance(h, 'hamming') * h.shape[0]\n",
    "\n",
    "    # Hierarchical clustering.\n",
    "    Z = sch.linkage(dist, method=linkage_method)\n",
    "\n",
    "        #harmonise metadata with samples\n",
    "    # Align sample metadata with haplotypes.\n",
    "\n",
    "    # Repeat the dataframe so there is one row of metadata for each haplotype.\n",
    "    leaf_data = pd.DataFrame(np.repeat(df_samples.values, 2, axis=0))\n",
    "    leaf_data.columns = df_samples.columns\n",
    "\n",
    "    sns.despine(ax=ax, offset=5, bottom=True, top=False)\n",
    "    # Compute the dendrogram but don't plot it.\n",
    "    dend = sch.dendrogram(\n",
    "        Z,\n",
    "        count_sort=count_sort,\n",
    "        distance_sort=distance_sort,\n",
    "        no_plot=False,\n",
    "        color_threshold=color_threshold,\n",
    "        above_threshold_color=above_threshold_color,\n",
    "    )\n",
    "\n",
    "    ax.get_xaxis().set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    sns.despine(ax=None, fig=None, offset=5, left=True, bottom=True, top=True, right=True)\n",
    "    ax.set_ylim(bottom=-2)\n",
    "    ax.set_ylabel('Distance (no. SNPs)')\n",
    "    ax.autoscale(axis='x', tight=True)\n",
    "\n",
    "    return Z, dend, leaf_data\n",
    "\n",
    "def fig_hap_structure(h, df_samples=df_samples, h_display=None, mutations=None, vspans=[[]], cluster_labels=[], figsize=(10, 8), \n",
    "                      fn=None, dpi=150, height_ratios=(3, .2, 1.5, .2), hap_pops=None, legend=False, title=None):\n",
    "    \n",
    "    # create the figure\n",
    "    fig = plt.figure(figsize=figsize)\n",
    "    \n",
    "    # define subplot layout\n",
    "    gs_nrows = 2\n",
    "    gs_ncols = 1\n",
    "    gs = mpl.gridspec.GridSpec(gs_nrows, gs_ncols, hspace=0.04, wspace=0.04,\n",
    "                               height_ratios=height_ratios)\n",
    "    \n",
    "    # dendrogram\n",
    "    ax_dend = fig.add_subplot(gs[0, 0])\n",
    "    z, r, leafdata = plot_dendrogram(ht, ax, df_samples)\n",
    "    ax_dend.set_ylim(bottom=-5)\n",
    "    if legend:\n",
    "        handles = [mpl.patches.Patch(color=analysis_popcols[pop], label=pop) for pop in analysis_popcols.keys()]\n",
    "        ax_dend.legend(handles=handles, loc='upper right', bbox_to_anchor=(1, 1), ncol=3)\n",
    "    ax_dend.set_yticklabels(ax_dend.get_yticks().astype(int))\n",
    "    ax_dend.set(xticklabels=[])\n",
    "    ax_dend.set_title(title,fontdict={'fontsize':18}, loc='right')\n",
    "\n",
    "    ax_dend.xaxis.set_tick_params(length=3, pad=2)\n",
    "    ax_dend.yaxis.set_tick_params(length=3, pad=2)\n",
    "\n",
    "    # population colours\n",
    "    ax_pops = fig.add_subplot(gs[1, 0])\n",
    "    if hap_pops is None:\n",
    "        hap_pops = leafdata.analysis_pop.values\n",
    "    x = hap_pops.take(r['leaves'])\n",
    "    hap_clrs = [analysis_popcols[p] for p in x]\n",
    "    ax_pops.broken_barh(xranges=[(i, 1) for i in range(h.shape[1])], yrange=(0, 1), color=hap_clrs);\n",
    "    sns.despine(ax=ax_pops, offset=2, left=True, bottom=True, top=True, right=True)\n",
    "    ax_pops.set_xticks([])\n",
    "    ax_pops.set_yticks([])\n",
    "    ax_pops.set_xlim(0, h.shape[1])\n",
    "    ax_pops.yaxis.set_label_position('left')\n",
    "    #ax_pops.set_ylabel('Population', rotation=0, ha='right', va='center')\n",
    "    fig.savefig(f'../figures/{title}.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [],
   "source": [
    "#dict of gene specific locations\n",
    "genes_main = {\n",
    "    'Ace1' : {'chrom' : 'CM023248', 'start' : 67473117, 'end': 67501071},\n",
    "    'Coeae' : {'chrom' : 'CM023249', 'start' : 18779698, 'end': 18804511},\n",
    "    'Gste' : {'chrom' : 'CM023249', 'start' : 70582788, 'end': 70594603},\n",
    "    'Cyp6' : {'chrom' : 'CM023248', 'start' : 67470071, 'end': 67514071},\n",
    "\n",
    "}\n",
    "\n",
    "genes_supp = {\n",
    "    'Vgsc' : {'chrom' : 'CM023249', 'start' : 42804885, 'end': 42848176},\n",
    "    'Rdl' : {'chrom' : 'CM023249', 'start' : 8345440, 'end': 8348441},\n",
    "    'Cyp9k1' : {'chrom' : 'CM023250', 'start' : 9713374, 'end': 9729212},\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now plot Genes for Main text\n",
    "\n",
    "fig, axs = plt.subplots(nrows=4, ncols=1, figsize=(6, 12))\n",
    "for i, gene in enumerate(genes_main.keys()):\n",
    "    haps = load_hap_ds(chrom = genes_main[gene]['chrom'],\n",
    "        start=genes_main[gene]['start'],\n",
    "        end=genes_main[gene]['end'],\n",
    "        min_minor_ac=1)\n",
    "\n",
    "    ht = allel.GenotypeArray(haps[1]['call_genotype']).to_haplotypes()\n",
    "    #iterate over genes and plot subplots\n",
    "    fig_hap_structure(h=ht, df_samples=df_samples, title=gene)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now plot Genes for Supp\n",
    "fig, axs = plt.subplots(nrows=4, ncols=1, figsize=(6, 12))\n",
    "for i, gene in enumerate(genes_supp.keys()):\n",
    "    haps = load_hap_ds(chrom = genes_supp[gene]['chrom'],\n",
    "        start=genes_supp[gene]['start'],\n",
    "        end=genes_supp[gene]['end'],\n",
    "        min_minor_ac=1)\n",
    "\n",
    "    ht = allel.GenotypeArray(haps[1]['call_genotype']).to_haplotypes()\n",
    "    #iterate over genes and plot subplots\n",
    "    fig_hap_structure(h=ht, df_samples=df_samples, title=gene)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ceasebasic",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
